---
title: "CWLã®ä½œæˆç’°å¢ƒã‚’VScodeã®ï½¢Dev Containersï½£ã®æ©Ÿèƒ½ã‚’ä½¿ã£ã¦æ§‹ç¯‰ã™ã‚‹ (3)"
emoji: "ğŸ“‘"
type: "tech" # tech: æŠ€è¡“è¨˜äº‹ / idea: ã‚¢ã‚¤ãƒ‡ã‚¢
topics: ["CWL", "bioinformatics"]
published: false
---
__â€»ä»Šå›ã®è¨˜äº‹ã§ä½¿ç”¨ã—ãŸCWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãŠã„ã¦ã„ã‚‹ãƒªãƒã‚¸ãƒˆãƒªã¯ä»¥ä¸‹ã‹ã‚‰ã‚¢ã‚¯ã‚»ã‚¹ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™__
https://github.com/yonezawa-sora/togotv_cwl_for_remote_container

:::message alert
ã“ã®èµ¤ã„ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã¯å¾Œã»ã©ä¿®æ­£ã™ã‚‹ç‚¹ã‚’ãƒ¡ãƒ¢ã—ã¦ã„ã¾ã™
:::

:::message alert
__ä¿®æ­£3ï¼šdockerã®èª¬æ˜ã®ã‚ã¨ã«singularityã®èª¬æ˜ã‚’åŠ ãˆã‚‹__
:::

# ãƒã‚¤ã‚ªã‚¤ãƒ³ãƒ•ã‚©ãƒãƒ†ã‚£ã‚¯ã‚¹ã®è§£ææ‰‹é †ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼åŒ–ã™ã‚‹

ã“ã‚Œã¾ã§ï½¤ç’°å¢ƒæ§‹ç¯‰ï½¤`grep`ã¨`wc`ã®å‡¦ç†ã«é–¢ã™ã‚‹ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’ä½œæˆã™ã‚‹éç¨‹ã‚’ç´¹ä»‹ã—ã¾ã—ãŸï½¡
ã“ã®ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã§ã¯ï½¤å…·ä½“çš„ã«ãƒã‚¤ã‚ªã‚¤ãƒ³ãƒ•ã‚©ãƒãƒ†ã‚£ã‚¯ã‚¹ç ”ç©¶ã§ä½¿ç”¨ã•ã‚Œã¦ã„ã‚‹ãƒ„ãƒ¼ãƒ«ã‚’ä½¿ã£ãŸä¾‹ã‚’cwlã¨ã—ã¦è¨˜è¿°ã—ï½¤ __ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼__ ã¨ã—ã¦å®Ÿè¡Œã™ã‚‹éƒ¨åˆ†ã¾ã§ã‚’ã”ç´¹ä»‹ã—ã¾ã™ï½¡
ã“ã®ãƒ—ãƒ­ã‚»ã‚¹ã«ãŠã„ã¦ã¯ï½¤zatsu-cwl-generatorã‚’ä½¿ã£ã¦CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½œæˆã—ï½¤ãã®å¾Œã«ç´°ã‹ã„ä¿®æ­£ã‚’åŠ ãˆã¦ã„ããªãŒã‚‰ä½œæˆã—ã¦ã„ãã¾ã™ï½¡

&nbsp;

ä»Šå›ã¯ä»¥ä¸‹ã«ç¤ºã™5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

**(1)blastpã‚³ãƒãƒ³ãƒ‰ã«ã‚ˆã‚‹é…åˆ—é¡ä¼¼æ€§æ¤œç´¢**
**(2)awkã«ã‚ˆã‚‹ãƒ’ãƒƒãƒˆã—ãŸIDã®æŠ½å‡º**
**(3)blastdbcmdã«ã‚ˆã‚‹ãƒ’ãƒƒãƒˆã—ãŸIDã®fastaãƒ•ã‚¡ã‚¤ãƒ«ã®æŠ½å‡º**
**(4)ClustalOmegaã«ã‚ˆã‚‹ãƒãƒ«ãƒãƒ—ãƒ«ã‚¢ãƒ©ã‚¤ãƒ³ãƒ¡ãƒ³ãƒˆ**
**(5)fasttreeã«ã‚ˆã‚‹newickå½¢å¼ãƒ•ã‚¡ã‚¤ãƒ«ã®å‡ºåŠ›**

ã“ã‚Œã«ã‚ˆã‚Šï½¤BLASTpã§ãƒ’ãƒƒãƒˆã—ãŸé…åˆ—ã‹ã‚‰ï½¤ç³»çµ±æ¨¹ã‚’ä½œæˆã™ã‚‹å‰æ®µéšã®å‡¦ç†(newickå½¢å¼ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å‡ºåŠ›)ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

:::message
äº‹å‰æº–å‚™ã§å¿…è¦ãªãƒ•ã‚¡ã‚¤ãƒ«ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ãªã©ã¯ä»¥ä¸‹ã®ãƒ—ãƒ­ã‚»ã‚¹ã§è¡Œã£ã¦ã„ã¾ã™ï½¡

:::details ãƒ•ã‚¡ã‚¤ãƒ«ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã¨ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã®ä½œæˆ

### 1\. ã‚¯ã‚¨ãƒªé…åˆ—

blastpã‚’å®Ÿè¡Œã™ã‚‹éš›ã«ã‚¯ã‚¨ãƒªã¨ã™ã‚‹ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãƒ‡ãƒ¼ã‚¿ã¨ã—ã¦ï½¤ã‚¦ã‚·ã®ãƒŸã‚ªã‚¹ã‚¿ãƒãƒ³ã®ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ã®fastaãƒ•ã‚¡ã‚¤ãƒ« (MSTN.fastaã«åå‰ã‚’å¤‰æ›´ã—ã¦ã„ã¾ã™)ã‚’ä½¿ç”¨ã—ã¾ã™ï½¡ 
curlã‚³ãƒãƒ³ãƒ‰ã§UniProtã‹ã‚‰å–å¾—ã—ã¾ã—ãŸï½¡

```bash:
curl -O https://rest.uniprot.org/uniprotkb/O18836.fasta
```

### 2\. ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹(ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹)

æ¬¡ã«ï½¤BLASTã®ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ã¨ã—ã¦ï½¤UniProtã®ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ã®fastaãƒ•ã‚¡ã‚¤ãƒ«(uniprot\_sprot.fasta.gz)ã‚’ftpã‚µã‚¤ãƒˆã‚ˆã‚Šcurlã‚³ãƒãƒ³ãƒ‰ã§å–å¾—ã—ï½¤å±•é–‹ã—ã¾ã™ï½¡Â 

```bash:
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
unpigz -p 10 uniprot_sprot.fasta.gz
``` 

### 3\. ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã®ä½œæˆ

ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã«ã™ã‚‹ãŸã‚ï½¤ `makeblastdb` ã‚³ãƒãƒ³ãƒ‰ã‚’dockerã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¾ã—ãŸ(æ—¢ã«ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«ã•ã‚Œã¦ã„ã‚‹å ´åˆã¯dockerã‚’ä½¿ç”¨ã—ãªãã¦ã‚‚å¤§ä¸ˆå¤«ã§ã™)ï½¡ 
`makeblastdb` ã‚³ãƒãƒ³ãƒ‰ ã®å®Ÿè¡Œã¯ä»¥ä¸‹ã®ã‚ˆã†ã«è¡Œã„ã¾ã—ãŸï½¡ 

```bash:
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 makeblastdb -in uniprot_sprot.fasta -dbtype prot -hash_index -parse_seqids 
```
:::

&nbsp;

CWLã¨ã—ã¦è¨˜è¿°ã™ã‚‹ã‚³ãƒãƒ³ãƒ‰ã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡ 
ã“ã®ã‚ˆã†ã«ï½¤5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

```bash:
# Step1
blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20 

# Step2
awk '{ print $2 }' blastp_result.txt > blastp_result_id.txt

#Step3
blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastdbcmd_result.fasta

#Step4
clustalo -i blastdbcmd_result.fasta --outfmt=fasta -o clustalo_result.fasta

#Step5
FastTree -boot 100 -out MSTN_tree.newick clustalo_result.fasta
```

:::message
ã“ã®è¨˜äº‹ã§ã‚‚å¾Œè¿°ã—ã¦ã„ã¾ã™ãŒï½¤__docker imageãŒã‚ã‚‹å ´åˆã«ã¯ãã®imageã‚’ä½¿ç”¨ã—ã¦å®Ÿè¡Œã™ã‚‹__ ã“ã¨ãŒã§ãã‚‹CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½œæˆã—ã¦ã„ãã®ã§ï½¤ã™ã§ã«ãƒ„ãƒ¼ãƒ«ã‚’ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«ã—ã¦ã„ã‚‹å ´åˆã§ã‚‚è©¦ã™ã“ã¨ãŒå¯èƒ½ã§ã™ï½¡
:::

&nbsp;

&nbsp;

## zatsu-cwl-generatorã‚’ä½¿ã£ã¦cwlãƒ•ã‚¡ã‚¤ãƒ«ã‚’æ›¸ã

æ¬¡ã«ï½¤ä¸Šè¨˜ã®å‡¦ç†ã«ã¤ã„ã¦ï½¤cwlãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¨˜è¿°ã—ã¦ã„ãã¾ã—ã‚‡ã†ï½¡ zatsu-cwl-generatorã‚’ä½¿ã£ã¦ã‚³ãƒ¼ãƒ‰ã‚’ç”Ÿæˆã—ï½¤ä¿®æ­£ã—ãªãŒã‚‰ä½œæˆã—ã¦ã„ãã¾ã™ï½¡
ä»Šå›ã¯ï½¤`zatsu_cwl_bioinformatics`ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã§ä½œæ¥­ã‚’è¡Œã£ã¦ã„ã¾ã™ï½¡ä¿®æ­£ã®ãƒ—ãƒ­ã‚»ã‚¹ã¯ãƒˆã‚°ãƒ«å†…ã«ç¤ºã—ã¦ã‚ã‚‹ã®ã§ï½¤ãœã²ã”è¦§ãã ã•ã„ï½¡
ä»¥ä¸‹ã§ã¯ï½¤æœ€åˆã®ãƒ—ãƒ­ã‚»ã‚¹ï½¤blastpã®ã‚³ãƒãƒ³ãƒ‰ã‚’ä¾‹ã«ã—ã¦(ä¿®æ­£ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚‚å«ã‚ãªãŒã‚‰)èª¬æ˜ã—ã¦ã„ãã¾ã™ï½¡

### (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦blastpã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã™ã‚‹

ã“ã‚Œã¾ã§ã¨åŒæ§˜ã«ï½¤zatsu-cwl-generatorã‚’ä½¿ã£ã¦CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã—ã¾ã™ï½¡

```bash:
zatsu-cwl-generator 'blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20' > 1_blastp.cwl
```

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp.cwl

`--validate`ã§ãƒã‚§ãƒƒã‚¯ã—ã¾ã™ï½¡
  
```bash:
cwltool --validate 1_blastp.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp.cwl'
1_blastp.cwl:45:7: Warning: Field 'location' contains undefined reference to
                   'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp.cwl:45:7: Warning: Field 'location' contains undefined reference to
                   'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
1_blastp.cwl is valid CWL.
```
ç”Ÿæˆã•ã‚ŒãŸãƒ•ã‚¡ã‚¤ãƒ«ã¯å•é¡ŒãŒãªã•ãã†ã§ã™ï½¡

:::message
ã“ã“ã§ä¸€æ—¦å®Ÿè¡Œã—ã¦ã¿ã‚ˆã†ã¨æ€ã„ã¾ã™ãŒï½¤ä¾‹ãˆã°ãƒ„ãƒ¼ãƒ«ã‚’ãƒ­ãƒ¼ã‚«ãƒ«ã§ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã—ã¦ãŠã‚‰ãšï½¤docker imageã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ãŸã„ã¨ã„ã†çŠ¶æ³ãŒã‚ã‚‹ã‹ã‚‚ã—ã‚Œã¾ã›ã‚“ï½¡
ä¾‹ãˆã°ï½¤ã“ã®è¨˜äº‹ã‚’ä½œæˆã—ã¦ã„ã‚‹ç’°å¢ƒã¯ï½¤Mac M1 ã§Dev containersã®æ©Ÿèƒ½ã‚’ä½¿ç”¨ã—ã¦ã„ã‚‹ãŸã‚ï½¤ç’°å¢ƒã¨ã—ã¦ã¯ __Linux ARM64__ ã¨ã„ã†çŠ¶æ³ã§ã™ï½¡
ç¾åœ¨ï½¤Linuxã®ARM64ã«å¯¾å¿œã—ãŸãƒ„ãƒ¼ãƒ«ã¯å°‘ãªãï½¤ã“ã®å ´åˆã§ã¯dockerã§å‹•ã‹ã—ãŸã»ã†ãŒç°¡å˜ãã†ã§ã™ï½¡
ã—ã‹ã—ãªãŒã‚‰ï½¤ __dockerã‚’ä½¿ã£ã¦å®Ÿè¡Œã™ã‚‹å ´åˆã«ã¯ï½¤ã©ã®ã‚ˆã†ã«cwlãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¨˜è¿°ã™ã‚Œã°ã‚ˆã„ã®ã§ã—ã‚‡ã†ã‹?__
ã“ã‚Œã«ã¤ã„ã¦ï½¤æ¬¡ã®ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã§èª¬æ˜ã—ã¦ã„ãã¾ã™ï½¡
:::

&nbsp;

### (2) ã‚³ãƒ³ãƒ†ãƒŠã‚ªãƒ—ã‚·ãƒ§ãƒ³(`-c`)ã‚’ä½¿ã£ã¦å‡ºåŠ›ã™ã‚‹

CWLã§ã¯ï½¤docker imageã‚’ä½¿ã£ã¦å®Ÿè¡Œã™ã‚‹å ´åˆï½¤ã“ã‚Œã¾ã§ã®cwlãƒ•ã‚¡ã‚¤ãƒ«ã§ã¯æ›¸ã„ã¦ã„ãªã‹ã£ãŸ`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰(ã‚ã‚‹ã„ã¯`requirements`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰)ã‚’ä½¿ã£ã¦è¨˜è¿°ã—ã¾ã™ï½¡

https://www.commonwl.org/user_guide/topics/using-containers.html#using-containers

__zatsu-cwl-generatorã§ã¯ï½¤`-c`, `--container` ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ä½¿ã†ã“ã¨ã§ï½¤ã‚³ãƒ³ãƒ†ãƒŠã‚’ä½¿ç”¨ã™ã‚‹ã‚³ãƒãƒ³ãƒ‰ã‚’å‡ºåŠ›ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡__

https://qiita.com/tm_tn/items/2c789c5b3c28e3eb3c9a

&nbsp;

ä»¥ä¸‹ã®ã‚ˆã†ã«ã‚³ãƒãƒ³ãƒ‰ã‚’å®Ÿè¡Œã—ã¾ã™ï½¡

```bash
zatsu-cwl-generator 'blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20' --container biocontainers/blast:v2.2.31_cv2 > 1_blastp_docker.cwl
```
ã“ã‚Œã¾ã§ã¨åŒã˜ã‚ˆã†ã«ã‚³ãƒãƒ³ãƒ‰ã‚’''ã§ããã£ãŸã‚ã¨ã«ï½¤`-c` ã‚‚ã—ãã¯`--container` ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã¦ï½¤image(tagã®æƒ…å ±ã‚‚å«ã‚ã¦)ã‚’è¨˜è¼‰ã—ã¾ã™ï½¡ 
ã™ã‚‹ã¨ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«CWLãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl

ã“ã‚Œã¾ã§ã¨åŒã˜ã‚ˆã†ã«å‡ºåŠ›ã•ã‚Œã¾ã™ãŒï½¤ã‚³ãƒ³ãƒ†ãƒŠã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã‚‹ã¨ï½¤ä¸‹ã«`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ãŒå‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl#L56-L58

ã“ã®ã‚ˆã†ã«ï½¤`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ä½¿ã£ã¦docker imageã‚’æŒ‡å®šã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

https://www.commonwl.org/v1.0/CommandLineTool.html#DockerRequirement

`--validate`ã‚’ä½¿ã£ã¦ãƒã‚§ãƒƒã‚¯ã—ã¾ã™ï½¡

```bash:
cwltool --validate 1_blastp_docker.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl'
1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
1_blastp_docker.cwl is valid CWL.
```
ä»Šã®ã¨ã“ã‚å•é¡Œã¯ãªã•ãã†ã§ã™ï½¡
ãã‚Œã§ã¯ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

```bash:
cwltool --debug 1_blastp_docker.cwl --query ./MSTN.fasta --db ./uniprot_sprot.fasta --num_threads 8 --outfmt 7 --out blastp_result.txt --max_target_seqs 20
```

ã“ã‚Œã¾ã§ã¨ç•°ãªã‚Šï½¤ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿã—ã¾ã—ãŸï½¡

:::details ã‚¨ãƒ©ãƒ¼å†…å®¹
```bash:
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl'
1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
    },
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl
DEBUG [job 1_blastp_docker.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
        "basename": "blastp_result.txt",
        "nameroot": "blastp_result",
        "nameext": ".txt"
    },
    "max_target_seqs": 20,
    "evalue": 1e-05
}
ERROR Input object failed validation:
[Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 178, in visit
    st = os.lstat(deref)
FileNotFoundError: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/cwltool/main.py", line 1314, in main
    (out, status) = real_executor(
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 63, in __call__
    return self.execute(process, job_order_object, runtime_context, logger)
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 146, in execute
    self.run_jobs(process, job_order_object, logger, runtime_context)
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 221, in run_jobs
    for job in jobiter:
  File "/usr/local/lib/python3.10/site-packages/cwltool/command_line_tool.py", line 992, in job
    builder.pathmapper = self.make_path_mapper(reffiles, builder.stagedir, runtimeContext, True)
  File "/usr/local/lib/python3.10/site-packages/cwltool/command_line_tool.py", line 485, in make_path_mapper
    return PathMapper(reffiles, runtimeContext.basedir, stagedir, separateDirs)
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 104, in __init__
    self.setup(dedup(referenced_files), basedir)
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 207, in setup
    self.visit(
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 167, in visit
    with SourceLine(
  File "schema_salad/sourceline.py", line 249, in __exit__
schema_salad.exceptions.ValidationException: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinforma
tics/blastp_result.txt'                                                                                                                            
```
:::

å‡ºåŠ›ãƒ•ã‚¡ã‚¤ãƒ«ãŒå­˜åœ¨ã—ãªã„ã¨ã„ã†ã‚¨ãƒ©ãƒ¼(`FileNotFoundError: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'`)ãŒå‡ºã¦ã„ã¾ã™ï½¡
ãã‚Œã§ã¯ã“ã‚Œã«ã¤ã„ã¦ä¿®æ­£ã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

&nbsp;

### (3) ä¿®æ­£ãƒ—ãƒ­ã‚»ã‚¹ 1

`inputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã«æ›¸ã„ã¦ã‚ã‚‹éƒ¨åˆ†ãŒä¿®æ­£ãŒå¿…è¦ã®ã‚ˆã†ã§ã™ï½¡
https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl#L41-L45

ã“ã“ã¯ï½¤ãƒ•ã‚¡ã‚¤ãƒ«ãã®ã‚‚ã®ã§ã¯ãªãï½¤__ãƒ•ã‚¡ã‚¤ãƒ«å__ ã«ãªã‚‹ã¯ãšãªã®ã§ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«ä¿®æ­£ã—ã¦ã¿ã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl#L41-L43

ãã‚Œã§ã¯å†åº¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡
:::details ä¿®æ­£å¾Œã®CWLãƒ•ã‚¡ã‚¤ãƒ«
```bash:1_blastp_docker_v2.cwl
cwltool --debug 1_blastp_docker_v2.cwl --query ./MSTN.fasta --db ./uniprot_sprot.fasta --num_threads 8 --outfmt 7 --out blastp_result.txt --max_target_seqs 20
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker_v2.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker_v2.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v2
.cwl                                                                                                                                               DEBUG [job 1_blastp_docker_v2.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20,
    "evalue": 1e-05
}
DEBUG [job 1_blastp_docker_v2.cwl] path mappings is {
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "/var/lib/cwl/stga7998393-9a81-4cc7-9591-8fe9236fcf5d/MSTN.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta",
        "File",
        true
    ]
}
DEBUG [job 1_blastp_docker_v2.cwl] command line bindings is [
    {
        "position": [
            -1000000,
            0
        ],
        "datum": "blastp"
    },
    {
        "position": [
            0,
            0
        ],
        "datum": "-query"
    },
    {
        "position": [
            0,
            1
        ],
        "valueFrom": "$(inputs.query)"
    },
    {
        "position": [
            0,
            2
        ],
        "datum": "-db"
    },
    {
        "position": [
            0,
            3
        ],
        "valueFrom": "$(inputs.db)"
    },
    {
        "position": [
            0,
            4
        ],
        "datum": "-evalue"
    },
    {
        "position": [
            0,
            5
        ],
        "valueFrom": "$(inputs.evalue)"
    },
    {
        "position": [
            0,
            6
        ],
        "datum": "-num_threads"
    },
    {
        "position": [
            0,
            7
        ],
        "valueFrom": "$(inputs.num_threads)"
    },
    {
        "position": [
            0,
            8
        ],
        "datum": "-outfmt"
    },
    {
        "position": [
            0,
            9
        ],
        "valueFrom": "$(inputs.outfmt)"
    },
    {
        "position": [
            0,
            10
        ],
        "datum": "-out"
    },
    {
        "position": [
            0,
            11
        ],
        "valueFrom": "$(inputs.out)"
    },
    {
        "position": [
            0,
            12
        ],
        "datum": "-max_target_seqs"
    },
    {
        "position": [
            0,
            13
        ],
        "valueFrom": "$(inputs.max_target_seqs)"
    }
]
DEBUG [job 1_blastp_docker_v2.cwl] initial work dir {}
INFO [job 1_blastp_docker_v2.cwl] /tmp/17begc9b$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/17begc9b,target=/jFvgAu \
    --mount=type=bind,source=/tmp/pvwugmew,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta,target=/var/lib/cwl/stga7998393-9a81-4
cc7-9591-8fe9236fcf5d/MSTN.fasta,readonly \                                                                                                            --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stgb5786c
95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta,readonly \                                                                                          --workdir=/jFvgAu \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/nquyx9tk/20240527074301-528880.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/jFvgAu \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stga7998393-9a81-4cc7-9591-8fe9236fcf5d/MSTN.fasta \
    -db \
    /var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta \
    -evalue \
    0.00001 \
    -num_threads \
    8 \
    -outfmt \
    7 \
    -out \
    blastp_result.txt \
    -max_target_seqs \
    20
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was reque
sted                                                                                                                                               BLAST Database error: No alias or index file found for protein database [/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta] 
in search path [/jFvgAu::]                                                                                                                         WARNING [job 1_blastp_docker_v2.cwl] exited with status: 2
WARNING [job 1_blastp_docker_v2.cwl] completed permanentFail
DEBUG [job 1_blastp_docker_v2.cwl] outputs {
    "all-for-debugging": [
        {
            "location": "file:///tmp/17begc9b/blastp_result.txt",
            "basename": "blastp_result.txt",
            "nameroot": "blastp_result",
            "nameext": ".txt",
            "class": "File",
            "checksum": "sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709",
            "size": 0,
            "http://commonwl.org/cwltool#generation": 0
        }
    ]
}
DEBUG [job 1_blastp_docker_v2.cwl] Removing input staging directory /tmp/e_qr0ldv
DEBUG [job 1_blastp_docker_v2.cwl] Removing temporary directory /tmp/pvwugmew
DEBUG Moving /tmp/17begc9b/blastp_result.txt to /workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt
DEBUG Removing intermediate output directory /tmp/17begc9b
{
    "all-for-debugging": [
        {
            "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
            "basename": "blastp_result.txt",
            "class": "File",
            "checksum": "sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709",
            "size": 0,
            "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
        }
    ]
}WARNING Final process status is permanentFail
```
:::
ä»Šå›ã‚‚ãƒ•ã‚¡ã‚¤ãƒ«ã®å‡ºåŠ›ã¯ã†ã¾ãã„ãã¾ã—ãŸãŒï½¤ç©ºã®ãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¦ã—ã¾ã„ã¾ã—ãŸï½¡
ã‚¨ãƒ©ãƒ¼ã‚’è¦‹ã¦ã¿ã‚‹ã¨ï½¤`BLAST Database error: No alias or index file found for protein database [/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta]`ãªã©ãŒè¦‹ã‚‰ã‚Œï½¤ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ãŒè¦‹ã¤ã‹ã‚‰ãªã„ã¨ã„ã†ã‚¨ãƒ©ãƒ¼ãŒå‡ºã¦ã„ã¾ã™ï½¡
ã“ã‚Œã¯BLASTãªã©ã«è¦‹ã‚‰ã‚Œã‚‹ã‚¨ãƒ©ãƒ¼ã§ï½¤ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ä½œæˆæ™‚ã«è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã‚‚å«ã‚ã¦è¨˜è¿°ã™ã‚‹å¿…è¦ãŒã‚ã‚Šã¾ã™ï½¡
ãã‚Œã§ã¯ï½¤ã“ã‚Œã‚’ä¿®æ­£ã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

&nbsp;

### (4) ä¿®æ­£ãƒ—ãƒ­ã‚»ã‚¹ 2 è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã®è¨˜è¼‰

`inputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã®éƒ¨åˆ†ã‚’ä¿®æ­£ã—ã¦ã¿ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl#L27-L31

BLASTã§ã¯ï½¤inputsã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹(å¼•æ•°ã¯ `-db` )ã®æŒ‡å®šã®éƒ¨åˆ†ã®è¨˜è¿°æ–¹æ³•ã«ã¤ã„ã¦ã„ãã¤ã‹æ–¹æ³•ãŒã‚ã‚‹ã‚ˆã†ã§ã™ï½¡

https://qiita.com/Yohei__K/items/b947655eb59a8853c172

ä»Šå›ã¯ `secondaryFiles` ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’æ´»ç”¨ã—ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«è¨˜è¿°ã—ã¦ã¿ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl#L27-L40

ã“ã®ã‚ˆã†ã«è¨˜è¿°ã™ã‚‹ã“ã¨ã§ï½¤è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã‚’æŒ‡å®šã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
ãã‚Œã§ã¯å†åº¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

:::details ä¿®æ­£å¾Œã®CWLãƒ•ã‚¡ã‚¤ãƒ«
```bash:1_blastp_docker_v3.cwl
cwltool --debug 1_blastp_docker_v3.cwl --query ./MSTN.fasta --db ./uniprot_sprot.fasta --num_threads 8 --outfmt 7 --out blastp_result.txt --max_target_seqs 20
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker_v3.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker_v3.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker_v3.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v3
.cwl                                                                                                                                               DEBUG [job 1_blastp_docker_v3.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta",
        "secondaryFiles": [
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd",
                "basename": "uniprot_sprot.fasta.phd",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phd"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi",
                "basename": "uniprot_sprot.fasta.phi",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phi"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr",
                "basename": "uniprot_sprot.fasta.phr",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phr"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin",
                "basename": "uniprot_sprot.fasta.pin",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".pin"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog",
                "basename": "uniprot_sprot.fasta.pog",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".pog"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd",
                "basename": "uniprot_sprot.fasta.psd",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psd"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi",
                "basename": "uniprot_sprot.fasta.psi",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psi"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq",
                "basename": "uniprot_sprot.fasta.psq",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psq"
            }
        ]
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20,
    "evalue": 1e-05
}
DEBUG [job 1_blastp_docker_v3.cwl] path mappings is {
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "/var/lib/cwl/stgcb307528-6f4f-4251-9517-0121cb068d7e/MSTN.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phd",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phi",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phr",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pin",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pog",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psd",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psi",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psq",
        "File",
        true
    ]
}
DEBUG [job 1_blastp_docker_v3.cwl] command line bindings is [
    {
        "position": [
            -1000000,
            0
        ],
        "datum": "blastp"
    },
    {
        "position": [
            0,
            0
        ],
        "datum": "-query"
    },
    {
        "position": [
            0,
            1
        ],
        "valueFrom": "$(inputs.query)"
    },
    {
        "position": [
            0,
            2
        ],
        "datum": "-db"
    },
    {
        "position": [
            0,
            3
        ],
        "valueFrom": "$(inputs.db)"
    },
    {
        "position": [
            0,
            4
        ],
        "datum": "-evalue"
    },
    {
        "position": [
            0,
            5
        ],
        "valueFrom": "$(inputs.evalue)"
    },
    {
        "position": [
            0,
            6
        ],
        "datum": "-num_threads"
    },
    {
        "position": [
            0,
            7
        ],
        "valueFrom": "$(inputs.num_threads)"
    },
    {
        "position": [
            0,
            8
        ],
        "datum": "-outfmt"
    },
    {
        "position": [
            0,
            9
        ],
        "valueFrom": "$(inputs.outfmt)"
    },
    {
        "position": [
            0,
            10
        ],
        "datum": "-out"
    },
    {
        "position": [
            0,
            11
        ],
        "valueFrom": "$(inputs.out)"
    },
    {
        "position": [
            0,
            12
        ],
        "datum": "-max_target_seqs"
    },
    {
        "position": [
            0,
            13
        ],
        "valueFrom": "$(inputs.max_target_seqs)"
    }
]
DEBUG [job 1_blastp_docker_v3.cwl] initial work dir {}
INFO [job 1_blastp_docker_v3.cwl] /tmp/8hge4mo3$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/8hge4mo3,target=/RrTBZF \
    --mount=type=bind,source=/tmp/7sot2rlv,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta,target=/var/lib/cwl/stgcb307528-6f4f-4
251-9517-0121cb068d7e/MSTN.fasta,readonly \                                                                                                            --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stg90d0f4
dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phd,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phi,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phr,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pin,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pog,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psd,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psi,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psq,readonly \                                                                                  --workdir=/RrTBZF \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/jj7sjx1j/20240527075910-603277.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/RrTBZF \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stgcb307528-6f4f-4251-9517-0121cb068d7e/MSTN.fasta \
    -db \
    /var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta \
    -evalue \
    0.00001 \
    -num_threads \
    8 \
    -outfmt \
    7 \
    -out \
    blastp_result.txt \
    -max_target_seqs \
    20
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was reque
sted                                                                                                                                               INFO [job 1_blastp_docker_v3.cwl] completed success
DEBUG [job 1_blastp_docker_v3.cwl] outputs {
    "all-for-debugging": [
        {
            "location": "file:///tmp/8hge4mo3/blastp_result.txt",
            "basename": "blastp_result.txt",
            "nameroot": "blastp_result",
            "nameext": ".txt",
            "class": "File",
            "checksum": "sha1$48540e3f3f63e4dd5a26ca6899cea47e8447849e",
            "size": 1923,
            "http://commonwl.org/cwltool#generation": 0
        }
    ]
}
DEBUG [job 1_blastp_docker_v3.cwl] Removing input staging directory /tmp/4qgo5l0s
DEBUG [job 1_blastp_docker_v3.cwl] Removing temporary directory /tmp/7sot2rlv
DEBUG Moving /tmp/8hge4mo3/blastp_result.txt to /workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt
DEBUG Removing intermediate output directory /tmp/8hge4mo3
{
    "all-for-debugging": [
        {
            "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
            "basename": "blastp_result.txt",
            "class": "File",
            "checksum": "sha1$48540e3f3f63e4dd5a26ca6899cea47e8447849e",
            "size": 1923,
            "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
        }
    ]
}INFO Final process status is success
```
:::

ä»Šå›ã¯æˆåŠŸã—ã¦ã„ã¾ã™! çµæœã‚’è¦‹ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blastp_result.txt

ã—ã£ã‹ã‚Šå‡ºåŠ›çµæœã‚‚å¾—ã‚‰ã‚Œã¦ã„ã¾ã™! ã‚¦ã‚·ã®ãƒŸã‚ªã‚¹ã‚¿ãƒãƒ³ã¨é…åˆ—é¡ä¼¼æ€§ãŒã‚ã‚‹ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãŒå¾—ã‚‰ã‚Œã¦ã„ã¾ã™ï½¡
ã“ã‚Œã§CWLã®ä¿®æ­£ã¯å®Œäº†ã§ã™ãŒï½¤æ›´ã«æ”¹å–„ã™ã‚‹ã“ã¨ã‚‚ã§ãã¾ã™ï½¡

:::message 

ä¸Šè¨˜ã®cwlãƒ•ã‚¡ã‚¤ãƒ«ã®ä¾‹ã§ã¯ï½¤E-valueã®æŒ‡å®šã‚’`Any`ã«ã—ã¦ã„ã¾ã—ãŸãŒï½¤å…·ä½“çš„ãªå‹ã‚’æŒ‡å®šã™ã‚‹ã“ã¨ã§ï½¤ã‚ˆã‚Šæ­£ç¢ºãªå…¥åŠ›ã‚’å—ã‘ä»˜ã‘ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl#L41-L43

`1e-05`ã®ã‚ˆã†ã«æŒ‡æ•°è¡¨è¨˜ã§å…¥åŠ›ã™ã‚‹ãŸã‚ï½¤`float?`ã®ã‚ˆã†ã«è¨˜è¼‰ã—ã¦ã¿ã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v4.cwl#L41-L43

ã“ã‚Œã«ã‚ˆã‚Šï½¤`Expecting one of: ['Directory', 'File', 'boolean', 'double', 'float', 'int', 'long', 'null', 'stderr', 'stdout', 'string']` ã¨ãƒã‚¤ãƒ©ã‚¤ãƒˆã•ã‚Œã¦ã„ãŸéƒ¨åˆ†ãŒè§£æ¶ˆã•ã‚Œã¾ã—ãŸï½¡
:::

****

&nbsp;

&nbsp;

&nbsp;

&nbsp;

## docker imageã‚’å–å¾—ã™ã‚‹

:::message
ã™ã§ã«æ§‹ç¯‰ã—ãŸç’°å¢ƒã«BLASTãªã©ã®ãƒ„ãƒ¼ãƒ«ã‚’ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«ã—ã¦ã„ã‚‹æ–¹ã¯ã“ã®æ‰‹é †ã¯é£›ã°ã—ã¦ãã ã•ã„ï½¡
:::

ä»Šå›å–ã‚Šä¸Šã’ã‚‹ãƒ„ãƒ¼ãƒ«ã®docker imageã‚’[docker hub](https://hub.docker.com/)ã‹ã‚‰å–å¾—ã—ã¾ã™ï½¡
[blast](https://hub.docker.com/r/biocontainers/blast)ï½¤[clustalo](https://hub.docker.com/r/biocontainers/clustalo)ï½¤[fasttree](https://hub.docker.com/r/biocontainers/fasttree)ã¯ã™ã¹ã¦[BioContainers](https://biocontainers.pro/)ã«ã‚ˆã£ã¦æ§‹ç¯‰ã•ã‚Œã¦ã„ã¾ã™[^1]ï½¡
ã¾ãŸï½¤awkã«ã¤ã„ã¦ã¯[ubuntu](https://hub.docker.com/_/ubuntu)ã®imageã‚’å–å¾—ã—ã¾ã—ãŸï½¡

:::message
ä»Šå›ã¯ï½¤tag ã‚’æŒ‡å®šã™ã‚‹å½¢ã§ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰(docker pull ... ã‚’ã‚³ãƒ”ãƒ¼&ãƒšãƒ¼ã‚¹ãƒˆ)ã—ã¾ã—ãŸï½¡
:::

ä»¥ä¸‹ã¯ã‚¿ãƒ¼ãƒŸãƒŠãƒ«ã§ `docker image ls` ã‚’è¡Œã£ãŸä¾‹ã§ã™ï½¡
```bash:
docker image ls
REPOSITORY               TAG                 IMAGE ID       CREATED       SIZE
ubuntu                   23.10               78938b354ee7   2 weeks ago   72.1MB
biocontainers/fasttree   v2.1.10-2-deb_cv1   bdfde5d6026c   4 years ago   118MB
biocontainers/clustalo   v1.2.4-2-deb_cv1    d14a5e27e301   4 years ago   118MB
biocontainers/blast      v2.2.31_cv2         5b25e08b9871   4 years ago   2.03GB
```
:

&nbsp;

&nbsp;

&nbsp;

```yaml
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result_MSTN.txt -max_target_seqs 20

class: CommandLineTool
cwlVersion: v1.0
baseCommand: [blastp] 

requirements:
  DockerRequirement: 
    dockerPull: "biocontainers/blast:v2.2.31_cv2"  #ã“ã“ã«ã‚¤ãƒ¡ãƒ¼ã‚¸ã‚’æ›¸ã
```
&nbsp; 

&nbsp;

&nbsp;

æ¬¡ã«Case1ã§ã‚‚æ›¸ã„ã¦ã„ãŸ`inputs` `outputs` ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã«ã¤ã„ã¦è¨˜è¿°ã—ã¾ã™ï½¡ `arguments`ã®éƒ¨åˆ†ã¯ä»Šå›ã¯è¨˜è¿°ã—ãªã„æ–¹æ³•ã§æ›¸ã„ã¦ã„ãã¾ã™ï½¡ å®Ÿéš›ã«æ›¸ã„ã¦ã¿ãŸã®ãŒä»¥ä¸‹ã®ã‚¹ã‚¯ãƒªãƒ—ãƒˆã«ãªã‚Šã¾ã™ï½¡

```yaml
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result_MSTN.txt -max_target_seqs 20
class: CommandLineTool
cwlVersion: v1.0
baseCommand: [blastp]
doc: BLASTP step 

requirements:
  DockerRequirement: #Dockerimageã‚’è¨˜è¿°
    dockerPull: "biocontainers/blast:v2.2.31_cv2"

# input
inputs:
  protein_query:
    type: File
    inputBinding:
      prefix: "-query"
      position: 1
  protein_database:
    type: File
    default: 
      class: File
      path: ../data/uniprot_sprot.fasta
    inputBinding:
      prefix: "-db"
      position: 2
    secondaryFiles:
      - ^.fasta.phd
      - ^.fasta.phi
      - ^.fasta.phr
      - ^.fasta.pin
      - ^.fasta.pog
      - ^.fasta.psd
      - ^.fasta.psi
      - ^.fasta.psq

  e-value:
    type: float?
    default: 1e-5
    inputBinding:
      prefix: "-evalue"
      position: 3
  number_of_threads:
    type: int
    default: 4
    inputBinding:
      prefix: "-num_threads"
      position: 4
  outformat_type:
    type: int
    default: 6
    inputBinding:
      prefix: "-outfmt"
      position: 5
  output_file_name:
    type: string
    inputBinding:
      prefix: "-out"
      position: 6
  max_target_sequence:
    type: int
    default: 20
    inputBinding:
      prefix: "-max_target_seqs"
      position: 7

#outputs
outputs:
  blastp_output_file:
    type: File
    outputBinding:
      glob: "$(inputs.output_file_name)"

# stdout
stdout: $(inputs.output_file_name)
```
&nbsp;

:::success
### BLASTã®ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã®æ›¸ãæ–¹

BLASTã§ã¯ï½¤inputsã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹(å¼•æ•°ã¯ `-db` )ã®æŒ‡å®šã®éƒ¨åˆ†ã®è¨˜è¿°æ–¹æ³•ã«ã¤ã„ã¦ã„ãã¤ã‹æ–¹æ³•ãŒã‚ã‚‹ã‚ˆã†ã§ã™ï½¡

ä¾‹ãˆã°ä»Šå›ã¯ `secondaryFiles` ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’æ´»ç”¨ã—ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«è¨˜è¿°ã—ã¦ã„ã¾ã™ï½¡ 

```yaml:
  protein_database:
    type: File
    default: 
      class: File
      path: ../data/uniprot_sprot.fasta #pathã‚’è¨˜è¿°
    inputBinding:
      prefix: "-db"
      position: 2
    secondaryFiles:
      - ^.fasta.phd
      - ^.fasta.phi
      - ^.fasta.phr
      - ^.fasta.pin
      - ^.fasta.pog
      - ^.fasta.psd
      - ^.fasta.psi
      - ^.fasta.psq
```
ãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦æŒ‡å®šã—ã¦ãŠãï½¤ãƒ‘ã‚¹ã«ã¤ã„ã¦ã‚‚æ˜è¨˜ã—ã¦ã„ã¾ã™ï½¡ã¾ãŸï½¤`secandaryFiles`ã§ï½¤ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹å‚ç…§æ™‚ã«protein_databaseã«æŒ‡å®šã—ã¦ã„ã‚‹ãƒ•ã‚¡ã‚¤ãƒ«ã¨åŒã˜ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã«ãªãã¦ã¯ã„ã‘ãªã„ãƒ•ã‚¡ã‚¤ãƒ«ã‚’æŒ‡å®šã—ã¦ã„ã¾ã™ï½¡

#### å‚è€ƒ1ï¼š[cwl-samples-2023/debuginfo.cwl](https://github.com/manabuishii/cwl-samples-2023/blob/main/debuginfo.cwl#L10-L11)
#### å‚è€ƒ2ï¼š[5.1 CommandInputParameter](https://www.commonwl.org/v1.0/CommandLineTool.html#CommandInputParameter)
:::
  
&nbsp;

### ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’YAMLãƒ•ã‚¡ã‚¤ãƒ«ã§ã¾ã¨ã‚ã¦ä¸ãˆã‚‹

ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’ä¸€ã¤ã®YAMLãƒ•ã‚¡ã‚¤ãƒ«ã«ã¾ã¨ã‚ã¦å®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
ãƒ•ã‚¡ã‚¤ãƒ«ã¯ä»¥ä¸‹ã®ã‚ˆã†ã«æ›¸ãã“ã¨ãŒå¯èƒ½ã§ã™ï½¡

```yaml:
protein_query: 
  class: File
  path: ../data/MSTN.fasta
protein_database: 
  class: File
  path: ../data/uniprot_sprot.fasta
e-value: 1e-5
number_of_threads: 4
outformat_type: 6
output_file_name: "blastp_result.txt"
max_target_sequence: 20

```
ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

&nbsp;

### 1_blastp.cwlã®å®Ÿè¡Œ

```bash=
cwltool --outdir ./data ./Tools/1_blastp.cwl ./Tools/1_blastp.yml
```
ä»Šå›ã¯`--outdir`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ä½¿ã£ã¦ï½¤å‡ºåŠ›å…ˆã®ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã‚’æŒ‡å®šã—ã¦å®Ÿè¡Œã—ã¦ã¿ã¾ã™ï½¡

&nbsp;

```bash:
INFO /usr/local/bin/cwltool 3.1.20230906142556
INFO Resolved './Tools/1_blastp.cwl' to 'file:///workspaces/togotv_shooting/Tools/1_blastp.cwl'
INFO [job 1_blastp.cwl] /tmp/qe5n7kgg$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/qe5n7kgg,target=/DcjUoi \
    --mount=type=bind,source=/tmp/p__s1km1,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-9100
606072ca/uniprot_sprot.fasta,readonly \                                                                                                  --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.phd,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.phd,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.phi,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.phi,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.phr,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.phr,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.pin,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.pin,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.pog,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.pog,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.psd,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.psd,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.psi,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.psi,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/uniprot_sprot.fasta.psq,target=/var/lib/cwl/stgd1ff326d-ac51-4352-b27b-
9100606072ca/uniprot_sprot.fasta.psq,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_shooting/data/MSTN.fasta,target=/var/lib/cwl/stge2ab10d6-20a1-4698-88fe-dc0997863178/
MSTN.fasta,readonly \                                                                                                                    --workdir=/DcjUoi \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/d3yca0_g/20230928025733-664451.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/DcjUoi \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stge2ab10d6-20a1-4698-88fe-dc0997863178/MSTN.fasta \
    -db \
    /var/lib/cwl/stgd1ff326d-ac51-4352-b27b-9100606072ca/uniprot_sprot.fasta \
    -evalue \
    0.00001 \
    -num_threads \
    4 \
    -outfmt \
    6 \
    -out \
    blastp_result.txt \
    -max_target_seqs \
    20
INFO [job 1_blastp.cwl] Max memory used: 11MiB
INFO [job 1_blastp.cwl] completed success
{
    "blastp_output_file": {
        "location": "file:///workspaces/togotv_shooting/data/blastp_result.txt",
        "basename": "blastp_result.txt",
        "class": "File",
        "checksum": "sha1$31464e2d0bdfd65f89e0542688844d3686bf278c",
        "size": 1536,
        "path": "/workspaces/togotv_shooting/data/blastp_result.txt"
    }
}INFO Final process status is success

```
å®Ÿè¡ŒãŒæˆåŠŸã—ãŸã‚ˆã†ã§ã™ï½¡blastpã®çµæœã‚‚å‡ºåŠ›ã•ã‚Œã¾ã—ãŸï½¡
åŒæ§˜ã«ä»–ã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚‚ï½¤zatsu-cwl-generatorã§ä¸€æ—¦å‡ºåŠ›å¾Œï½¤ä¿®æ­£ã™ã‚‹å½¢ã§æ›¸ã„ã¦ã¿ã¾ã—ã‚‡ã†ï½¡ãã®çµæœã‚’ä»¥ä¸‹ã«ç¤ºã—ã¾ã™ï½¡ 

&nbsp;

## CWLãƒ•ã‚¡ã‚¤ãƒ«(CommandLineTool)è¨˜è¿°ä¾‹

### 2_awk.cwl

```yaml:
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v `pwd`:`pwd` -w `pwd` ubuntu:23.10 awk '{ print $2 }' blastp_result_MSTN.txt > blastp_result_id.txt
class: CommandLineTool
cwlVersion: v1.0
baseCommand: [awk]

requirements:
  DockerRequirement:
    dockerPull: "ubuntu:23.10"

# å›ºå®šã®å¼•æ•°
arguments:
  - valueFrom: '{ print $2 }' 
    position: 1

inputs:
  blastp_result:
    type: File
    inputBinding:
      position: 2
  output_id_file_name:
    type: string
    inputBinding:
      position: 3
outputs:
  awk_output:
    type: File
    outputBinding:
      glob: "$(inputs.output_id_file_name)"
stdout: $(inputs.output_id_file_name)
```

&nbsp;

### 3\_blastdbcmd.cwl

```yaml:
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v /workspaces/togotv_shooting/data:/workspaces/togotv_shooting/data  -w  /workspaces/togotv_shooting/data biocontainers/blast:v2.2.31_cv2 blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastp_results_MSTN.fasta
class: CommandLineTool
cwlVersion: v1.0
baseCommand: [blastdbcmd]

requirements:
  DockerRequirement:
    dockerPull: "biocontainers/blast:v2.2.31_cv2"

# Inputs
inputs:
  blastdbcmd_protein_database:
    type: File
    default: 
      class: File
      path: ../data/uniprot_sprot.fasta
    inputBinding:
      prefix: "-db"
      position: 1
    secondaryFiles:
      - ^.fasta.phd
      - ^.fasta.phi
      - ^.fasta.phr
      - ^.fasta.pin
      - ^.fasta.pog
      - ^.fasta.psd
      - ^.fasta.psi
      - ^.fasta.psq
  id_query:
    type: File
    inputBinding:
      prefix: "-entry_batch"
      position: 2
  blastdbcmd_output_name:
    type: string
    inputBinding:
      prefix: "-out"
      position: 3
#Outputs
outputs:
  blastdbcmd_output:
    type: File
    outputBinding:
      glob: "$(inputs.blastdbcmd_output_name)"
```

&nbsp;

### 4_clustalo.cwl

```yaml:
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v /workspaces/togotv_shooting/data:/workspaces/togotv_shooting/data -w /workspaces/togotv_shooting/data biocontainers/clustalo:v1.2.4-2-deb_cv1 clustalo -i /workspaces/togotv_shooting/data/blastp_results_MSTN.fasta --outfmt=fasta -o MSTN_aligned_sequence.fasta
class: CommandLineTool
cwlVersion: v1.0
baseCommand: [clustalo]

requirements:
  DockerRequirement:
    dockerPull: "biocontainers/clustalo:v1.2.4-2-deb_cv1"

# Inputs
inputs:
  clustalo_inputs:
    type: File
    inputBinding:
      prefix: "-i"
      position: 1
  clustalo_format:
    type: string
    default: "fasta"
    inputBinding:
      prefix: "--outfmt"
      position: 2
  clustalo_output_name:
    type: string
    inputBinding:
      prefix: "-o"
      position: 3

outputs:
  clustalo_output:
    type: File
    outputBinding:
      glob: "$(inputs.clustalo_output_name)"
```

&nbsp; 

### 5_fasttree.cwl

```yaml:
#!/usr/bin/env cwl-runner
# Generated from: docker run --rm -it -v /workspaces/togotv_shooting/data:/workspaces/togotv_shooting/data -w /workspaces/togotv_shooting/data biocontainers/fasttree:v2.1.10-2-deb_cv1 FastTree -out MSTN_tree.newick /workspaces/togotv_shooting/data/MSTN_aligned_sequence.fasta
class: CommandLineTool
cwlVersion: v1.0
baseCommand: [fasttree]

requirements:
  DockerRequirement:
    dockerPull: "biocontainers/fasttree:v2.1.10-2-deb_cv1"

#Inputs
inputs:
  fasttree_output_file_name:
    type: string
    inputBinding:
      prefix: "-out"
      position: 1
  fasttree_input_file:
    type: File
    inputBinding:
      position: 2
# Outputs
outputs:
  fasttree_output_file:
    type: File
    outputBinding:
      glob: "$(inputs.fasttree_output_file_name)" 
```

&nbsp;

### ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’è¨˜è¿°ã™ã‚‹

:::danger
__ä¿®æ­£ï¼šbest practicesã«è¼‰ã£ã¦ã„ã‚‹ã‚ˆã†ãªæ›¸ãæ–¹ã§æœ€å¾Œæ›¸ã__
:::

ã“ã‚Œã¾ã§ï½¤Commanlinetoolã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’æ›¸ã„ã¦ã¿ã¾ã—ãŸï½¡æ¬¡ã«ã“ã‚Œã‚‰ã®5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’å®Ÿè¡Œã™ã‚‹ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’è¨˜è¿°ã—ã¦ã„ãã¾ã™ï½¡ã“ã®ä¾‹ã§ã¯ï½¤ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼å…¨ä½“ã«é–¢ã™ã‚‹ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’ï½¢1\_protein\_queryï½£ã®ã‚ˆã†ã«æ•°å­—ã‚’ã¤ã‘ã¦ã„ã¾ã™ï½¡ å‰ã®å‡¦ç†ã®ã‚¢ã‚¦ãƒˆãƒ—ãƒƒãƒˆã‚’å—ã‘å–ã‚‹éƒ¨åˆ†ã¯ï½¢blastp\_result: step1\_blastp/blastp\_output\_fileï½£ã®ã‚ˆã†ã«è¨˜è¿°ã—ï½¤ãã‚Œä»¥å¤–ã®å…¨ã¦ã®å‡¦ç†ã«é–¢ã™ã‚‹ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’inputsã§è¨˜è¿°ã—ã¦ã„ã¾ã™ï½¡
```yaml:
#!/usr/bin/env cwl-runner
cwlVersion: v1.0
class: Workflow
doc: "5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’å®Ÿè¡Œã™ã‚‹ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼: blastp, awk, blastdbcmd, clustalo, and fasttree"
#æœ€åˆã®ã‚¤ãƒ³ãƒ—ãƒƒãƒˆã®ã¿è¨˜è¿°ã—ã‚ˆã†ã¨æ€ã£ãŸãŒï½¤å…¨éƒ¨æ›¸ã„ã¦ã‚ã‹ã‚Šã‚„ã™ãæ›¸ãæ–¹é‡ã«
# å‰ã®ã‚¹ãƒ†ãƒƒãƒ—ã®å‡ºåŠ›ã®æ›¸ãæ–¹ã«æ³¨æ„
inputs: 
  #blastp
  1_protein_query: 
    type: File
  2_protein_database:
    type: File
    secondaryFiles:
      - ^.fasta.phd
      - ^.fasta.phi
      - ^.fasta.phr
      - ^.fasta.pin
      - ^.fasta.pog
      - ^.fasta.psd
      - ^.fasta.psi
      - ^.fasta.psq
  3_e-value: 
    type: float?
  4_number_of_threads: 
    type: int
  5_outformat_type: 
    type: int
  6_output_file_name: 
    type: string
  7_max_target_sequence: 
    type: int
  #awk
  8_output_id_file_name:
    type: string
  #blastdbcmd
  9_blastdbcmd_protein_database:
    type: File
    secondaryFiles:
      - ^.fasta.phd
      - ^.fasta.phi
      - ^.fasta.phr
      - ^.fasta.pin
      - ^.fasta.pog
      - ^.fasta.psd
      - ^.fasta.psi
      - ^.fasta.psq
  10_blastdbcmd_output_name:
    type: string
  #clustalo
  11_clustalo_format:
    type: string
  12_clustalo_output_name:
    type: string
  #fasttree
  13_fasttree_output_file_name:
    type: string
  

outputs:
  final_output_bootstrap: 
    type: File
    outputSource: step5_fasttree/fasttree_output_file
steps:
  step1_blastp:
    run: ../Tools/1_blastp.cwl 
    in: 
      protein_query: 1_protein_query
      protein_database: 2_protein_database
      e-value: 3_e-value
      number_of_threads: 4_number_of_threads
      outformat_type: 5_outformat_type
      output_file_name: 6_output_file_name
      max_target_sequence: 7_max_target_sequence
    out: [blastp_output_file] 

  step2_awk:
    run: ../Tools/2_awk.cwl
    in:
      blastp_result: step1_blastp/blastp_output_file 
      output_id_file_name: 8_output_id_file_name
    out: [awk_output]

  step3_blastdbcmd:
    run: ../Tools/3_blastdbcmd.cwl
    in: 
      blastdbcmd_protein_database: 9_blastdbcmd_protein_database
      id_query: step2_awk/awk_output 
      blastdbcmd_output_name: 10_blastdbcmd_output_name
    out: [blastdbcmd_output]

  step4_clustalo:
    run: ../Tools/4_clustalo.cwl
    in:
      clustalo_inputs: step3_blastdbcmd/blastdbcmd_output
      clustalo_format: 11_clustalo_format
      clustalo_output_name: 12_clustalo_output_name
    out: [clustalo_output]

  step5_fasttree:
    run: ../Tools/5_fasttree.cwl
    in:
      fasttree_output_file_name: 13_fasttree_output_file_name
      fasttree_input_file: step4_clustalo/clustalo_output
    out: [fasttree_output_file]
```

æœ€å¾Œã«å…¨ã¦ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿(1\_protein\_queryã®ã‚ˆã†ã«æ›¸ã„ã¦ã„ã‚‹éƒ¨åˆ†)ã‚’è¨˜è¿°ã—ãŸãƒ•ã‚¡ã‚¤ãƒ« ã‚’ä½œæˆã—ã¾ã™ï½¡ã™ã§ã«è¿°ã¹ãŸã‚ˆã†ã«ï½¤ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã¯YAMLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦ã¾ã¨ã‚ã¦è¨˜è¿°ã™ã‚‹ã“ã¨ãŒå¯èƒ½ã§ã™ï½¡

ä»¥ä¸‹ã®ã‚ˆã†ãªè¨˜è¿°ã«ãªã‚Šã¾ã—ãŸï½¡

&nbsp;

### ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ãƒ•ã‚¡ã‚¤ãƒ«(YAMLãƒ•ã‚¡ã‚¤ãƒ«)

```yaml:
# 1_blastp.cwl ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿
1_protein_query:
  class: File
  path: ../data/MSTN.fasta
2_protein_database:
  class: File
  path: ../data/uniprot_sprot.fasta
3_e-value: 1e-5
4_number_of_threads: 4
5_outformat_type: 6
6_output_file_name: "blastp_result.txt"
7_max_target_sequence: 20

# 2_awk.cwl ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿
8_output_id_file_name: "blastp_result_id.txt"

# 3_blastdbcmd.cwl ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿
9_blastdbcmd_protein_database:
  class: File
  path: "../data/uniprot_sprot.fasta"
10_blastdbcmd_output_name: "blastdbcmd_output.fasta"

# 4_clustalo.cwl ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿
11_clustalo_format: "fasta"
12_clustalo_output_name: "clustalo_output.fasta"

# 5_fasttree.cwl ã®ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿
13_fasttree_output_file_name: "fasttree_output.newick"
```

&nbsp;

æœ€å¾Œã«`cwltool â€“-validate`ã‚’å®Ÿè¡Œã—ã¾ã™ï½¡
```yaml:
cwltool --validate ./workflow/6_phylogenetic-workflow.cwl ./workflow/input.yaml
INFO /usr/local/bin/cwltool 3.1.20230906142556
INFO Resolved './workflow/6_phylogenetic-workflow.cwl' to 'file:///workspaces/togotv_shooting/workflow/6_phylogenetic-workflow.cwl'
./workflow/6_phylogenetic-workflow.cwl is valid CWL.
```

æ›¸ãæ–¹ã¨ã—ã¦ã¯å•é¡ŒãŒç„¡ã„ã‚ˆã†ã§ã™ï½¡
ç¶šã„ã¦ï½¤CWLãƒ•ã‚¡ã‚¤ãƒ«ã«å¯¾ã—ã¦`--help`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’è¡Œã£ã¦ã¿ã¾ã™ï½¡

```bash:
cwltool ./workflow/6_phylogenetic-workflow.cwl --help
```

:::danger
__è¿½è¨˜ï¼šã“ã“ã§ã‚‚å…ˆé ­ä½•è¡Œã‹è¦‹ã›ã¦ï½¤docãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’æ›¸ãã“ã¨ã®ã”åˆ©ç›ŠãŒå¾—ã‚‰ã‚Œã‚‹ã“ã¨ã‚’æ›¸ã__
:::


ãã‚Œã§ã¯ï½¤å®Ÿéš›ã«å®Ÿè¡Œã—ã¦ã¿ã¾ã™ï½¡

```yaml:
cwl-runner ./workflow/6_phylogenetic-workflow.cwl ./workflow/input.yml
```

å®Ÿè¡Œçµæœ
```bash:

```

è§£æãŒæˆåŠŸã—ã¦ã„ã‚‹ã‚ˆã†ã§ã™ï½¡ å®Ÿéš›ã«ï½¤newickå½¢å¼ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚‚å‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡ ã“ã®ã‚ˆã†ã«ï½¤ä½•å›ã‚‚ã‚³ãƒãƒ³ãƒ‰ã‚’å®Ÿè¡Œã›ãšï½¤1ã¤ã®ã‚³ãƒãƒ³ãƒ‰ã§å®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
__CWLã§è¨˜è¿°ã—ãŸãƒ•ã‚¡ã‚¤ãƒ«ã¯æ§˜ã€…ãªç’°å¢ƒã§å®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ã®ã§ï½¤ãœã²ã”è‡ªèº«ã§ã‚¯ãƒ­ãƒ¼ãƒ³ã—ã¦ç¢ºã‹ã‚ã¦ãã ã•ã„!__

&nbsp;

:::details å‚è€ƒï¼šDockerã‚³ãƒãƒ³ãƒ‰ã§å®Ÿè¡Œã™ã‚‹å ´åˆ(ç¢ºèªæ¸ˆã¿)
```bash:
# Step1
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result_MSTN.txt -max_target_seqs 20 

# Step2
docker run --rm -it -v `pwd`:`pwd` -w `pwd` ubuntu:23.10 awk '{ print $2 }' `pwd`/data/blastp_result.txt > `pwd`/data/blastp_result_id.txt

#Step3
docker run --rm -it -v `pwd`:`pwd`  -w  `pwd` biocontainers/blast:v2.2.31_cv2 blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastdbcmd_result.fasta

#Step4
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/clustalo:v1.2.4-2-deb_cv1 clustalo -i `pwd`/blastp_results_MSTN.fasta --outfmt=fasta -o clustalo_result.fasta

#Step5
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/fasttree:v2.1.10-2-deb_cv1 FastTree -boot 100 -out MSTN_tree.newick `pwd`/clustalo_result.fasta
```
:::

[^1]: [ãƒã‚¤ã‚ªã‚¤ãƒ³ãƒ•ã‚©ãƒãƒ†ã‚£ã‚¯ã‚¹ã®ãƒ„ãƒ¼ãƒ«ã‚’å†ç¾æ€§ã‚ˆãå®Ÿè¡Œã™ã‚‹ãŸã‚ã®ã‚³ãƒ³ãƒ†ãƒŠä»®æƒ³åŒ–ãƒ„ãƒ¼ãƒ«ç¾¤ BioContainers](https://kazumaxneo.hatenablog.com/entry/2018/10/02/112232)