---
title: "CWLã®ä½œæˆç’°å¢ƒã‚’VScodeã®ï½¢Dev Containersï½£ã®æ©Ÿèƒ½ã‚’ä½¿ã£ã¦æ§‹ç¯‰ã™ã‚‹ (3)"
emoji: "ğŸ“‘"
type: "tech" # tech: æŠ€è¡“è¨˜äº‹ / idea: ã‚¢ã‚¤ãƒ‡ã‚¢
topics: ["CWL", "bioinformatics"]
published: true
---
__ä»Šå›ã®è¨˜äº‹ã§ä½¿ç”¨ã—ãŸCWLã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãŠã„ã¦ã„ã‚‹ãƒªãƒã‚¸ãƒˆãƒªã¯ä»¥ä¸‹ã‹ã‚‰ã‚¢ã‚¯ã‚»ã‚¹ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡__
https://github.com/yonesora56/togotv_cwl_for_remote_container

å‰å›ã®è¨˜äº‹ã¯ã“ã¡ã‚‰ã§ã™
https://zenn.dev/sorayone/articles/cwl-document_1

https://zenn.dev/sorayone/articles/cwl-document_2

:::message
__æœ¬è¨˜äº‹ã®å¯¾è±¡ã¨ãªã‚‹æ–¹__
â–  ç’°å¢ƒæ§‹ç¯‰ãŒã§ããŸã®ã§ï½¤æ¬¡ã®ã‚¹ãƒ†ãƒƒãƒ—ã«é€²ã¿ãŸã„æ–¹
â–  å®Ÿéš›ã®è§£æã®å ´é¢ã§CWLã‚’ä½¿ã£ã¦ã¿ãŸã„æ–¹

(1)ï½¤(2)ã®è¨˜äº‹ã‚’è¦‹ã¦ã„ãŸã ã„ãŸã‚ã¨ï½¤æ›´ã«CWLã‚’å‹‰å¼·ã—ãŸã„ã¨ã„ã†æ–¹å‘ã‘ã®è¨˜äº‹ã«ãªã‚Šã¾ã™ï½¡
:::

# ã¯ã˜ã‚ã«

ã“ã‚Œã¾ã§ï½¤ç’°å¢ƒæ§‹ç¯‰ãŠã‚ˆã³`grep`ã¨`wc`ã®å‡¦ç†ã«é–¢ã™ã‚‹ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’ä½œæˆã™ã‚‹éç¨‹ã‚’ç´¹ä»‹ã—ã¾ã—ãŸï½¡
ã“ã®è¨˜äº‹ã§ã¯ï½¤å…·ä½“çš„ã«ãƒã‚¤ã‚ªã‚¤ãƒ³ãƒ•ã‚©ãƒãƒ†ã‚£ã‚¯ã‚¹åˆ†é‡ã§ä½¿ç”¨ã•ã‚Œã¦ã„ã‚‹ãƒ„ãƒ¼ãƒ«ã®ä¾‹ã‚’CWLã¨ã—ã¦è¨˜è¿°ã—ï½¤ __ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼__ ã¨ã—ã¦å®Ÿè¡Œã™ã‚‹éƒ¨åˆ†ã¾ã§ã‚’ã”ç´¹ä»‹ã—ã¾ã™ï½¡
ã“ã®ãƒ—ãƒ­ã‚»ã‚¹ã«ãŠã„ã¦ã¯ï½¤ã“ã‚Œã¾ã§ã¨åŒæ§˜ã«zatsu-cwl-generatorã‚’è»¸ã¨ã—ã¦CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½œæˆã—ï½¤ãã®å¾Œã«ç´°ã‹ã„ä¿®æ­£ã‚’åŠ ãˆã¦ã„ããªãŒã‚‰ä½œæˆã—ã¦ã„ãã¾ã™ï½¡

## å®Ÿè¡Œã™ã‚‹ãƒ—ãƒ­ã‚»ã‚¹ã®æ¦‚è¦

ä»Šå›ã¯ä»¥ä¸‹ã«ç¤ºã™5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

**(1)blastpã‚³ãƒãƒ³ãƒ‰ã«ã‚ˆã‚‹é…åˆ—é¡ä¼¼æ€§æ¤œç´¢**
**(2)awkã«ã‚ˆã‚‹ãƒ’ãƒƒãƒˆã—ãŸIDã®æŠ½å‡º**
**(3)blastdbcmdã«ã‚ˆã‚‹ãƒ’ãƒƒãƒˆã—ãŸIDã®fastaãƒ•ã‚¡ã‚¤ãƒ«ã®æŠ½å‡º**
**(4)ClustalOmegaã«ã‚ˆã‚‹ãƒãƒ«ãƒãƒ—ãƒ«ã‚¢ãƒ©ã‚¤ãƒ³ãƒ¡ãƒ³ãƒˆ**
**(5)fasttreeã«ã‚ˆã‚‹newickå½¢å¼ãƒ•ã‚¡ã‚¤ãƒ«ã®å‡ºåŠ›**

ã“ã‚Œã«ã‚ˆã‚Šï½¤BLASTpã§ãƒ’ãƒƒãƒˆã—ãŸé…åˆ—ã‹ã‚‰ï½¤ç³»çµ±æ¨¹ã‚’ä½œæˆã™ã‚‹å‰æ®µéšã®å‡¦ç†(newickå½¢å¼ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å‡ºåŠ›)ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

:::message
äº‹å‰æº–å‚™ã§å¿…è¦ãªãƒ•ã‚¡ã‚¤ãƒ«ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ãªã©ã¯ä»¥ä¸‹ã®ãƒ—ãƒ­ã‚»ã‚¹ã§è¡Œã£ã¦ã„ã¾ã™ï½¡
:::details ãƒ•ã‚¡ã‚¤ãƒ«ã®ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã¨ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã®ä½œæˆ
### 1\. ã‚¯ã‚¨ãƒªé…åˆ—ã®å–å¾—

blastpã‚’å®Ÿè¡Œã™ã‚‹éš›ã«ã‚¯ã‚¨ãƒªã¨ã™ã‚‹ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãƒ‡ãƒ¼ã‚¿ã¨ã—ã¦ï½¤ã‚¦ã‚·ã®ãƒŸã‚ªã‚¹ã‚¿ãƒãƒ³ã®ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ã®fastaãƒ•ã‚¡ã‚¤ãƒ« (MSTN.fastaã«åå‰ã‚’å¤‰æ›´ã—ã¦ã„ã¾ã™)ã‚’ä½¿ç”¨ã—ã¾ã™ï½¡ 
curlã‚³ãƒãƒ³ãƒ‰ã§UniProtã‹ã‚‰å–å¾—ã—ã¾ã—ãŸï½¡

```bash:
curl -O https://rest.uniprot.org/uniprotkb/O18836.fasta
```

### 2\. uniprot fastaãƒ•ã‚¡ã‚¤ãƒ«ã®å–å¾—

æ¬¡ã«ï½¤BLASTã®ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ã¨ã—ã¦ï½¤UniProtã®ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ã®fastaãƒ•ã‚¡ã‚¤ãƒ«(uniprot\_sprot.fasta.gz)ã‚’ftpã‚µã‚¤ãƒˆã‚ˆã‚Šcurlã‚³ãƒãƒ³ãƒ‰ã§å–å¾—ã—ï½¤å±•é–‹ã—ã¾ã™ï½¡Â 

```bash:
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
pigz -d -p 10 uniprot_sprot.fasta.gz
``` 

### 3\. ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã®ä½œæˆ

ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ã«ã™ã‚‹ãŸã‚ï½¤ `makeblastdb` ã‚³ãƒãƒ³ãƒ‰ã‚’dockerã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¾ã—ãŸ(æ—¢ã«ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«ã•ã‚Œã¦ã„ã‚‹å ´åˆã¯dockerã‚’ä½¿ç”¨ã—ãªãã¦ã‚‚å¤§ä¸ˆå¤«ã§ã™)ï½¡ 
`makeblastdb` ã‚³ãƒãƒ³ãƒ‰ ã®å®Ÿè¡Œã¯ä»¥ä¸‹ã®ã‚ˆã†ã«è¡Œã„ã¾ã—ãŸï½¡ 

```bash:
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 makeblastdb -in uniprot_sprot.fasta -dbtype prot -hash_index -parse_seqids 
```
:::

&nbsp;

CWLã¨ã—ã¦è¨˜è¿°ã™ã‚‹ã‚³ãƒãƒ³ãƒ‰ã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡ 
ã“ã®ã‚ˆã†ã«ï½¤5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¿°ã—ã¾ã™ï½¡

```bash:
# Step1
blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20 

# Step2
awk '{ print $2 }' blastp_result.txt > blastp_result_id.txt

#Step3
blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastdbcmd_result.fasta

#Step4
clustalo -i blastdbcmd_result.fasta --outfmt=fasta -o clustalo_result.fasta

#Step5
FastTree -boot 100 -out MSTN_tree.newick clustalo_result.fasta
```

:::message
ã“ã®è¨˜äº‹ã§ã‚‚å¾Œè¿°ã—ã¦ã„ã¾ã™ãŒï½¤__docker imageãŒã‚ã‚‹å ´åˆã«ã¯ãã®imageã‚’ä½¿ç”¨ã—ã¦å®Ÿè¡Œã™ã‚‹__ ã“ã¨ãŒã§ãã‚‹CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½œæˆã—ã¦ã„ãã®ã§ï½¤ã™ã§ã«ãƒ„ãƒ¼ãƒ«ã‚’ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«ã—ã¦ã„ã‚‹å ´åˆã§ã‚‚è©¦ã™ã“ã¨ãŒå¯èƒ½ã§ã™ï½¡
:::

&nbsp;

# zatsu-cwl-generatorã‚’ä½¿ã£ã¦blastpã‚’å®Ÿè¡Œã™ã‚‹CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’æ›¸ã

æ¬¡ã«ï½¤ä¸Šè¨˜ã®å‡¦ç†ã«ã¤ã„ã¦ï½¤CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¨˜è¿°ã—ã¦ã„ãã¾ã—ã‚‡ã†ï½¡ zatsu-cwl-generatorã‚’ä½¿ã£ã¦ã‚³ãƒ¼ãƒ‰ã‚’ç”Ÿæˆã—ï½¤ä¿®æ­£ã—ãªãŒã‚‰ä½œæˆã—ã¦ã„ãã¾ã™ï½¡
ä»Šå›ã¯ï½¤`zatsu_cwl_bioinformatics`ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã§ä½œæ¥­ã‚’è¡Œã£ã¦ã„ã¾ã™ï½¡ä¿®æ­£ã®ãƒ—ãƒ­ã‚»ã‚¹ã¯ãƒˆã‚°ãƒ«å†…ã«ç¤ºã—ã¦ã‚ã‚‹ã®ã§ï½¤ãœã²ã”è¦§ãã ã•ã„ï½¡
ä»¥ä¸‹ã§ã¯ï½¤æœ€åˆã®ãƒ—ãƒ­ã‚»ã‚¹ï½¤blastpã®ã‚³ãƒãƒ³ãƒ‰ã‚’ä¾‹ã«ã—ã¦(ä¿®æ­£ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚‚å«ã‚ãªãŒã‚‰)èª¬æ˜ã—ã¦ã„ãã¾ã™ï½¡

## (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦blastpã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã™ã‚‹

ã“ã‚Œã¾ã§ã¨åŒæ§˜ã«ï½¤zatsu-cwl-generatorã‚’ä½¿ã£ã¦CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã—ã¾ã™ï½¡

```bash:
zatsu-cwl-generator 'blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20' > 1_blastp.cwl
```

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp.cwl

`cwltool --validate`ã§ãƒã‚§ãƒƒã‚¯ã—ã¾ã™ï½¡
  
```bash:
cwltool --validate 1_blastp.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp.cwl'
1_blastp.cwl:45:7: Warning: Field 'location' contains undefined reference to
                   'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp.cwl:45:7: Warning: Field 'location' contains undefined reference to
                   'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
1_blastp.cwl is valid CWL.
```
`1_blastp.cwl is valid CWL.`ã¨ã„ã†å‡ºåŠ›ãŒã•ã‚Œã¦ã„ã‚‹ã‚‚ã®ã®ï½¤`1_blastp.cwl is valid CWL.`ã¨ã—ã¦å®Ÿè¡Œä¾‹ã¨ã—ã¦ä½¿ã£ãŸå…¥åŠ›ãƒ•ã‚¡ã‚¤ãƒ«ãŒå­˜åœ¨ã—ãªã„ã“ã¨ã‚’è¡¨ã™è­¦å‘ŠãŒå‡ºã¦ã„ã¾ã™ãŒã€ä¿®æ­£æ–¹æ³•ã¯å¾Œã»ã©èª¬æ˜ã—ã¾ã™ï½¡

:::message
ã“ã“ã§ä¸€æ—¦å®Ÿè¡Œã—ã¦ã¿ã‚ˆã†ã¨æ€ã„ã¾ã™ãŒï½¤ä¾‹ãˆã°ãƒ„ãƒ¼ãƒ«ã‚’ãƒ­ãƒ¼ã‚«ãƒ«ã§ãƒ€ã‚¦ãƒ³ãƒ­ãƒ¼ãƒ‰ã—ã¦ãŠã‚‰ãšï½¤docker imageã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ãŸã„ã¨ã„ã†çŠ¶æ³ãŒã‚ã‚‹ã‹ã‚‚ã—ã‚Œã¾ã›ã‚“ï½¡ 
ä¾‹ãˆã°ï½¤ã“ã®è¨˜äº‹ã‚’ä½œæˆã—ã¦ã„ã‚‹ç’°å¢ƒã¯ï½¤Mac M1 ã§Dev containersã®æ©Ÿèƒ½ã‚’ä½¿ç”¨ã—ã¦ã„ã‚‹ãŸã‚ï½¤ç’°å¢ƒã¨ã—ã¦ã¯ __Linux ARM64__ ã¨ã„ã†çŠ¶æ³ã§ã™ï½¡ ç¾åœ¨ï½¤Linuxã®ARM64ã«å¯¾å¿œã—ãŸãƒ„ãƒ¼ãƒ«ã¯å°‘ãªãï½¤ã“ã®å ´åˆã§ã¯dockerã§å‹•ã‹ã—ãŸã»ã†ãŒç°¡å˜ãã†ã§ã™ï½¡
ã—ã‹ã—ãªãŒã‚‰ï½¤ __dockerã‚’ä½¿ã£ã¦å®Ÿè¡Œã™ã‚‹å ´åˆã«ã¯ï½¤ã©ã®ã‚ˆã†ã«CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¨˜è¿°ã™ã‚Œã°ã‚ˆã„ã®ã§ã—ã‚‡ã†ã‹?__
ã“ã‚Œã«ã¤ã„ã¦ï½¤æ¬¡ã®ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã§èª¬æ˜ã—ã¦ã„ãã¾ã™ï½¡
:::

&nbsp;

## (2) ã‚³ãƒ³ãƒ†ãƒŠã‚ªãƒ—ã‚·ãƒ§ãƒ³(`-c`)ã‚’ä½¿ã£ã¦å‡ºåŠ›ã™ã‚‹

CWLã§ã¯ï½¤docker imageã‚’ä½¿ã£ã¦å®Ÿè¡Œã™ã‚‹å ´åˆï½¤ã“ã‚Œã¾ã§ã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã§ã¯æ›¸ã„ã¦ã„ãªã‹ã£ãŸ`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰(ã‚ã‚‹ã„ã¯`requirements`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰)ã‚’ä½¿ã£ã¦è¨˜è¿°ã—ã¾ã™ï½¡

https://www.commonwl.org/user_guide/topics/using-containers.html#using-containers

__zatsu-cwl-generatorã§ã¯ï½¤`-c`, `--container` ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ä½¿ã†ã“ã¨ã§ï½¤ã‚³ãƒ³ãƒ†ãƒŠã‚’ä½¿ç”¨ã™ã‚‹ã‚³ãƒãƒ³ãƒ‰ã‚’å‡ºåŠ›ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡__

https://qiita.com/tm_tn/items/2c789c5b3c28e3eb3c9a

&nbsp;

ä»¥ä¸‹ã®ã‚ˆã†ã«ã‚³ãƒãƒ³ãƒ‰ã‚’å®Ÿè¡Œã—ã¾ã™ï½¡

```bash
zatsu-cwl-generator 'blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result.txt -max_target_seqs 20' --container biocontainers/blast:v2.2.31_cv2 > 1_blastp_docker.cwl
```
ã“ã‚Œã¾ã§ã¨åŒã˜ã‚ˆã†ã«ã‚³ãƒãƒ³ãƒ‰ã‚’''ã§ããã£ãŸã‚ã¨ã«ï½¤`-c` ã‚‚ã—ãã¯`--container` ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã¦ï½¤image(tagã®æƒ…å ±ã‚‚å«ã‚ã¦)ã‚’è¨˜è¼‰ã—ã¾ã™ï½¡ ã™ã‚‹ã¨ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«CWLãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl

ã“ã‚Œã¾ã§ã¨åŒã˜ã‚ˆã†ã«å‡ºåŠ›ã•ã‚Œã¾ã™ãŒï½¤ã‚³ãƒ³ãƒ†ãƒŠã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã‚‹ã¨ï½¤ä¸‹ã«`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ãŒå‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl#L56-L58

ã“ã®ã‚ˆã†ã«ï½¤`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ä½¿ã£ã¦docker imageã‚’æŒ‡å®šã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

https://www.commonwl.org/v1.0/CommandLineTool.html#DockerRequirement

`cwltool --validate`ã‚’ä½¿ã£ã¦ãƒã‚§ãƒƒã‚¯ã—ã¾ã™ï½¡

```bash:
cwltool --validate 1_blastp_docker.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl'
1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
1_blastp_docker.cwl is valid CWL.
```
ã“ã¡ã‚‰ã«ã¤ã„ã¦ã‚‚ï½¤`1_blastp_docker.cwl is valid CWL.`ã¨ã„ã†å‡ºåŠ›ãŒã•ã‚Œã¦ã„ã‚‹ã‚‚ã®ã®ï½¤`1_blastp_docker.cwl is valid CWL.`ã¨ã—ã¦å®Ÿè¡Œä¾‹ã¨ã—ã¦ä½¿ã£ãŸå…¥åŠ›ãƒ•ã‚¡ã‚¤ãƒ«ãŒå­˜åœ¨ã—ãªã„ã“ã¨ã‚’è¡¨ã™è­¦å‘ŠãŒå‡ºã¦ã„ã¾ã™ãŒï½¤ä¿®æ­£æ–¹æ³•ã¯å¾Œã»ã©èª¬æ˜ã—ã¾ã™ï½¡
ä¸€æ—¦ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

```bash:
cwltool --debug 1_blastp_docker.cwl
```

ã“ã‚Œã¾ã§ã¨ç•°ãªã‚Šï½¤ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿã—ã¾ã—ãŸï½¡

:::details ã‚¨ãƒ©ãƒ¼å†…å®¹
```bash:
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl'
1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
WARNING 1_blastp_docker.cwl:45:7: Warning: Field 'location' contains undefined reference to
                          'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
    },
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker.cwl
DEBUG [job 1_blastp_docker.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
        "basename": "blastp_result.txt",
        "nameroot": "blastp_result",
        "nameext": ".txt"
    },
    "max_target_seqs": 20,
    "evalue": 1e-05
}
ERROR Input object failed validation:
[Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 178, in visit
    st = os.lstat(deref)
FileNotFoundError: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/cwltool/main.py", line 1314, in main
    (out, status) = real_executor(
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 63, in __call__
    return self.execute(process, job_order_object, runtime_context, logger)
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 146, in execute
    self.run_jobs(process, job_order_object, logger, runtime_context)
  File "/usr/local/lib/python3.10/site-packages/cwltool/executors.py", line 221, in run_jobs
    for job in jobiter:
  File "/usr/local/lib/python3.10/site-packages/cwltool/command_line_tool.py", line 992, in job
    builder.pathmapper = self.make_path_mapper(reffiles, builder.stagedir, runtimeContext, True)
  File "/usr/local/lib/python3.10/site-packages/cwltool/command_line_tool.py", line 485, in make_path_mapper
    return PathMapper(reffiles, runtimeContext.basedir, stagedir, separateDirs)
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 104, in __init__
    self.setup(dedup(referenced_files), basedir)
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 207, in setup
    self.visit(
  File "/usr/local/lib/python3.10/site-packages/cwltool/pathmapper.py", line 167, in visit
    with SourceLine(
  File "schema_salad/sourceline.py", line 249, in __exit__
schema_salad.exceptions.ValidationException: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinforma
tics/blastp_result.txt'                                                                                                                            
```
:::

å‡ºåŠ›ãƒ•ã‚¡ã‚¤ãƒ«ãŒå­˜åœ¨ã—ãªã„ã¨ã„ã†ã‚¨ãƒ©ãƒ¼(`FileNotFoundError: [Errno 2] No such file or directory: '/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt'`)ãŒå‡ºã¦ã„ã¾ã™ï½¡
ã“ã‚Œã¯`cwltool --validate`ã‚’å®Ÿè¡Œã—ãŸã¨ãã®è­¦å‘Šã¨é–¢ä¿‚ãŒã‚ã‚‹ã‚ˆã†ã§ã™ï½¡
ãã‚Œã§ã¯ã“ã‚Œã«ã¤ã„ã¦ä¿®æ­£ã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

&nbsp;

## (3) ä¿®æ­£ãƒ—ãƒ­ã‚»ã‚¹ 1

`inputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã«æ›¸ã„ã¦ã‚ã‚‹éƒ¨åˆ†ãŒä¿®æ­£ãŒå¿…è¦ã®ã‚ˆã†ã§ã™ï½¡
https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker.cwl#L41-L45

ã“ã“ã®`blastp_result.txt`ã¯ãƒ•ã‚¡ã‚¤ãƒ«ãã®ã‚‚ã®ã§ã¯ãªãï½¤__ãƒ•ã‚¡ã‚¤ãƒ«å__ ï½¤ã¤ã¾ã‚Šæ–‡å­—åˆ—ã«ãªã‚‹ã¯ãšãªã®ã§ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«ä¿®æ­£ã—ã¦ã¿ã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl#L41-L43

ãã‚Œã§ã¯å†åº¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡
:::details ä¿®æ­£å¾Œã®CWLãƒ•ã‚¡ã‚¤ãƒ«
```bash:1_blastp_docker_v2.cwl
cwltool --debug 1_blastp_docker_v2.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker_v2.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker_v2.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v2
.cwl                                                                                                                                               DEBUG [job 1_blastp_docker_v2.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20,
    "evalue": 1e-05
}
DEBUG [job 1_blastp_docker_v2.cwl] path mappings is {
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "/var/lib/cwl/stga7998393-9a81-4cc7-9591-8fe9236fcf5d/MSTN.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta",
        "File",
        true
    ]
}
DEBUG [job 1_blastp_docker_v2.cwl] command line bindings is [
    {
        "position": [
            -1000000,
            0
        ],
        "datum": "blastp"
    },
    {
        "position": [
            0,
            0
        ],
        "datum": "-query"
    },
    {
        "position": [
            0,
            1
        ],
        "valueFrom": "$(inputs.query)"
    },
    {
        "position": [
            0,
            2
        ],
        "datum": "-db"
    },
    {
        "position": [
            0,
            3
        ],
        "valueFrom": "$(inputs.db)"
    },
    {
        "position": [
            0,
            4
        ],
        "datum": "-evalue"
    },
    {
        "position": [
            0,
            5
        ],
        "valueFrom": "$(inputs.evalue)"
    },
    {
        "position": [
            0,
            6
        ],
        "datum": "-num_threads"
    },
    {
        "position": [
            0,
            7
        ],
        "valueFrom": "$(inputs.num_threads)"
    },
    {
        "position": [
            0,
            8
        ],
        "datum": "-outfmt"
    },
    {
        "position": [
            0,
            9
        ],
        "valueFrom": "$(inputs.outfmt)"
    },
    {
        "position": [
            0,
            10
        ],
        "datum": "-out"
    },
    {
        "position": [
            0,
            11
        ],
        "valueFrom": "$(inputs.out)"
    },
    {
        "position": [
            0,
            12
        ],
        "datum": "-max_target_seqs"
    },
    {
        "position": [
            0,
            13
        ],
        "valueFrom": "$(inputs.max_target_seqs)"
    }
]
DEBUG [job 1_blastp_docker_v2.cwl] initial work dir {}
INFO [job 1_blastp_docker_v2.cwl] /tmp/17begc9b$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/17begc9b,target=/jFvgAu \
    --mount=type=bind,source=/tmp/pvwugmew,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta,target=/var/lib/cwl/stga7998393-9a81-4
cc7-9591-8fe9236fcf5d/MSTN.fasta,readonly \                                                                                                            --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stgb5786c
95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta,readonly \                                                                                          --workdir=/jFvgAu \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/nquyx9tk/20240527074301-528880.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/jFvgAu \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stga7998393-9a81-4cc7-9591-8fe9236fcf5d/MSTN.fasta \
    -db \
    /var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta \
    -evalue \
    0.00001 \
    -num_threads \
    8 \
    -outfmt \
    7 \
    -out \
    blastp_result.txt \
    -max_target_seqs \
    20
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was reque
sted                                                                                                                                               BLAST Database error: No alias or index file found for protein database [/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta] 
in search path [/jFvgAu::]                                                                                                                         WARNING [job 1_blastp_docker_v2.cwl] exited with status: 2
WARNING [job 1_blastp_docker_v2.cwl] completed permanentFail
DEBUG [job 1_blastp_docker_v2.cwl] outputs {
    "all-for-debugging": [
        {
            "location": "file:///tmp/17begc9b/blastp_result.txt",
            "basename": "blastp_result.txt",
            "nameroot": "blastp_result",
            "nameext": ".txt",
            "class": "File",
            "checksum": "sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709",
            "size": 0,
            "http://commonwl.org/cwltool#generation": 0
        }
    ]
}
DEBUG [job 1_blastp_docker_v2.cwl] Removing input staging directory /tmp/e_qr0ldv
DEBUG [job 1_blastp_docker_v2.cwl] Removing temporary directory /tmp/pvwugmew
DEBUG Moving /tmp/17begc9b/blastp_result.txt to /workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt
DEBUG Removing intermediate output directory /tmp/17begc9b
{
    "all-for-debugging": [
        {
            "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
            "basename": "blastp_result.txt",
            "class": "File",
            "checksum": "sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709",
            "size": 0,
            "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
        }
    ]
}WARNING Final process status is permanentFail
```
:::
ä»Šå›ã¯ãƒ•ã‚¡ã‚¤ãƒ«ã®å‡ºåŠ›ã¯ã•ã‚ŒãŸã‚‚ã®ã®ï½¤ç©ºã®ãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¦ã—ã¾ã„ã¾ã—ãŸï½¡
ã‚¨ãƒ©ãƒ¼ã‚’è¦‹ã¦ã¿ã‚‹ã¨ï½¤`BLAST Database error: No alias or index file found for protein database [/var/lib/cwl/stgb5786c95-6694-4f59-9707-d79eced66e64/uniprot_sprot.fasta]`ãªã©ãŒè¦‹ã‚‰ã‚Œï½¤ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ãŒè¦‹ã¤ã‹ã‚‰ãªã„ã¨ã„ã†ã‚¨ãƒ©ãƒ¼ãŒå‡ºã¦ã„ã¾ã™ï½¡
ã“ã‚Œã¯BLASTãªã©ã«è¦‹ã‚‰ã‚Œã‚‹ã‚¨ãƒ©ãƒ¼ã§ï½¤ __ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ä½œæˆæ™‚ã«è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã‚‚å«ã‚ã¦è¨˜è¿°ã™ã‚‹å¿…è¦ãŒã‚ã‚Šã¾ã™ï½¡__
ãã‚Œã§ã¯ï½¤ã“ã‚Œã‚’ä¿®æ­£ã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

&nbsp;

## (4) ä¿®æ­£ãƒ—ãƒ­ã‚»ã‚¹ 2 è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã®è¨˜è¼‰

`inputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã®éƒ¨åˆ†ã‚’ä¿®æ­£ã—ã¦ã¿ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v2.cwl#L27-L31

BLASTã§ã¯ï½¤inputsã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹(å¼•æ•°ã¯ `-db` )ã®æŒ‡å®šã®éƒ¨åˆ†ã®è¨˜è¿°æ–¹æ³•ã«ã¤ã„ã¦ä¸‹è¨˜ã®ã‚ˆã†ãªæ–¹æ³•ãŒã‚ã‚‹ã‚ˆã†ã§ã™ï½¡

https://qiita.com/Yohei__K/items/b947655eb59a8853c172

ã“ã®è¨˜äº‹ã§ã¯åˆ¥ã®ã‚¢ãƒ—ãƒ­ãƒ¼ãƒã¨ã—ã¦ï½¤ `secondaryFiles` ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’æ´»ç”¨ã—ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«è¨˜è¿°ã—ã¦ã¿ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl#L27-L40

ã“ã®ã‚ˆã†ã«è¨˜è¿°ã™ã‚‹ã“ã¨ã§ï½¤è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã‚’æŒ‡å®šã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

https://www.commonwl.org/v1.0/CommandLineTool.html#CommandInputParameter

ãã‚Œã§ã¯å†åº¦å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

:::details ä¿®æ­£å¾Œã®cwlãƒ•ã‚¡ã‚¤ãƒ«ã§ã®å®Ÿè¡Œçµæœ
```bash:1_blastp_docker_v3.cwl
cwltool --debug 1_blastp_docker_v3.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '1_blastp_docker_v3.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl'
DEBUG Can't make command line argument from Any
DEBUG Parsed job order from command line: {
    "__id": "1_blastp_docker_v3.cwl",
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta"
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20
}
DEBUG [job 1_blastp_docker_v3.cwl] initializing from file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/1_blastp_docker_v3
.cwl                                                                                                                                               DEBUG [job 1_blastp_docker_v3.cwl] {
    "query": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "size": 476,
        "basename": "MSTN.fasta",
        "nameroot": "MSTN",
        "nameext": ".fasta"
    },
    "db": {
        "class": "File",
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "size": 284812502,
        "basename": "uniprot_sprot.fasta",
        "nameroot": "uniprot_sprot",
        "nameext": ".fasta",
        "secondaryFiles": [
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd",
                "basename": "uniprot_sprot.fasta.phd",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phd"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi",
                "basename": "uniprot_sprot.fasta.phi",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phi"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr",
                "basename": "uniprot_sprot.fasta.phr",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".phr"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin",
                "basename": "uniprot_sprot.fasta.pin",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".pin"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog",
                "basename": "uniprot_sprot.fasta.pog",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".pog"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd",
                "basename": "uniprot_sprot.fasta.psd",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psd"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi",
                "basename": "uniprot_sprot.fasta.psi",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psi"
            },
            {
                "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq",
                "basename": "uniprot_sprot.fasta.psq",
                "class": "File",
                "nameroot": "uniprot_sprot.fasta",
                "nameext": ".psq"
            }
        ]
    },
    "num_threads": 8,
    "outfmt": 7,
    "out": "blastp_result.txt",
    "max_target_seqs": 20,
    "evalue": 1e-05
}
DEBUG [job 1_blastp_docker_v3.cwl] path mappings is {
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta",
        "/var/lib/cwl/stgcb307528-6f4f-4251-9517-0121cb068d7e/MSTN.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phd",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phi",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phr",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pin",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pog",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psd",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psi",
        "File",
        true
    ],
    "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq": [
        "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq",
        "/var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psq",
        "File",
        true
    ]
}
DEBUG [job 1_blastp_docker_v3.cwl] command line bindings is [
    {
        "position": [
            -1000000,
            0
        ],
        "datum": "blastp"
    },
    {
        "position": [
            0,
            0
        ],
        "datum": "-query"
    },
    {
        "position": [
            0,
            1
        ],
        "valueFrom": "$(inputs.query)"
    },
    {
        "position": [
            0,
            2
        ],
        "datum": "-db"
    },
    {
        "position": [
            0,
            3
        ],
        "valueFrom": "$(inputs.db)"
    },
    {
        "position": [
            0,
            4
        ],
        "datum": "-evalue"
    },
    {
        "position": [
            0,
            5
        ],
        "valueFrom": "$(inputs.evalue)"
    },
    {
        "position": [
            0,
            6
        ],
        "datum": "-num_threads"
    },
    {
        "position": [
            0,
            7
        ],
        "valueFrom": "$(inputs.num_threads)"
    },
    {
        "position": [
            0,
            8
        ],
        "datum": "-outfmt"
    },
    {
        "position": [
            0,
            9
        ],
        "valueFrom": "$(inputs.outfmt)"
    },
    {
        "position": [
            0,
            10
        ],
        "datum": "-out"
    },
    {
        "position": [
            0,
            11
        ],
        "valueFrom": "$(inputs.out)"
    },
    {
        "position": [
            0,
            12
        ],
        "datum": "-max_target_seqs"
    },
    {
        "position": [
            0,
            13
        ],
        "valueFrom": "$(inputs.max_target_seqs)"
    }
]
DEBUG [job 1_blastp_docker_v3.cwl] initial work dir {}
INFO [job 1_blastp_docker_v3.cwl] /tmp/8hge4mo3$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/8hge4mo3,target=/RrTBZF \
    --mount=type=bind,source=/tmp/7sot2rlv,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta,target=/var/lib/cwl/stgcb307528-6f4f-4
251-9517-0121cb068d7e/MSTN.fasta,readonly \                                                                                                            --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stg90d0f4
dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta,readonly \                                                                                          --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phd,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phi,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.phr,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pin,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.pog,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psd,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psi,readonly \                                                                                  --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq,target=/var/lib/cwl/stg90
d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta.psq,readonly \                                                                                  --workdir=/RrTBZF \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/jj7sjx1j/20240527075910-603277.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/RrTBZF \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stgcb307528-6f4f-4251-9517-0121cb068d7e/MSTN.fasta \
    -db \
    /var/lib/cwl/stg90d0f4dc-8137-4082-8c5d-4eaf670f8786/uniprot_sprot.fasta \
    -evalue \
    0.00001 \
    -num_threads \
    8 \
    -outfmt \
    7 \
    -out \
    blastp_result.txt \
    -max_target_seqs \
    20
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was reque
sted                                                                                                                                               INFO [job 1_blastp_docker_v3.cwl] completed success
DEBUG [job 1_blastp_docker_v3.cwl] outputs {
    "all-for-debugging": [
        {
            "location": "file:///tmp/8hge4mo3/blastp_result.txt",
            "basename": "blastp_result.txt",
            "nameroot": "blastp_result",
            "nameext": ".txt",
            "class": "File",
            "checksum": "sha1$48540e3f3f63e4dd5a26ca6899cea47e8447849e",
            "size": 1923,
            "http://commonwl.org/cwltool#generation": 0
        }
    ]
}
DEBUG [job 1_blastp_docker_v3.cwl] Removing input staging directory /tmp/4qgo5l0s
DEBUG [job 1_blastp_docker_v3.cwl] Removing temporary directory /tmp/7sot2rlv
DEBUG Moving /tmp/8hge4mo3/blastp_result.txt to /workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt
DEBUG Removing intermediate output directory /tmp/8hge4mo3
{
    "all-for-debugging": [
        {
            "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt",
            "basename": "blastp_result.txt",
            "class": "File",
            "checksum": "sha1$48540e3f3f63e4dd5a26ca6899cea47e8447849e",
            "size": 1923,
            "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blastp_result.txt"
        }
    ]
}INFO Final process status is success
```
:::

ä»Šå›ã¯æˆåŠŸã—ã¦ã„ã¾ã™! çµæœã‚’è¦‹ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blastp_result.txt

ã—ã£ã‹ã‚Šå‡ºåŠ›çµæœã‚‚å¾—ã‚‰ã‚Œã¦ã„ã¾ã™! ã‚¦ã‚·ã®ãƒŸã‚ªã‚¹ã‚¿ãƒãƒ³ã¨é…åˆ—é¡ä¼¼æ€§ãŒã‚ã‚‹ã‚¿ãƒ³ãƒ‘ã‚¯è³ªé…åˆ—ãŒå¾—ã‚‰ã‚Œã¦ã„ã¾ã™ï½¡
ã“ã‚Œã§CWLã®ä¿®æ­£ã¯å®Œäº†ã§ã™ãŒï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«ï½¤æ›´ã«æ”¹å–„ã™ã‚‹ã“ã¨ã‚‚ã§ãã¾ã™ï½¡

## (5) å‹ã®æŒ‡å®š

ä¸Šè¨˜ã®cwlãƒ•ã‚¡ã‚¤ãƒ«ã®ä¾‹ã§ã¯ï½¤E-valueã®æŒ‡å®šã‚’`Any`ã«ã—ã¦ã„ã¾ã—ãŸãŒï½¤å…·ä½“çš„ãªå‹ã‚’æŒ‡å®šã™ã‚‹ã“ã¨ã§ï½¤ã‚ˆã‚Šæ­£ç¢ºãªå…¥åŠ›ã‚’å—ã‘ä»˜ã‘ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v3.cwl#L41-L43

E-valueã¯`1e-05`ã®ã‚ˆã†ã«å…¥åŠ›ã™ã‚‹ãŸã‚ï½¤`float`ã¨ã—ã¦è¨˜è¼‰ã—ã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v4.cwl#L41-L43

ã“ã‚Œã«ã‚ˆã‚Šï½¤`Expecting one of: ['Directory', 'File', 'boolean', 'double', 'float', 'int', 'long', 'null', 'stderr', 'stdout', 'string']` ã¨ãƒã‚¤ãƒ©ã‚¤ãƒˆã•ã‚Œã¦ã„ãŸéƒ¨åˆ†ãŒè§£æ¶ˆã•ã‚Œã¾ã—ãŸï½¡
(ã“ã®éƒ¨åˆ†ã‚’ä¿®æ­£ã—ãŸãƒ•ã‚¡ã‚¤ãƒ«ã¯`v4`ã¨ã—ã¦ä¿å­˜ã—ã¦ã„ã¾ã™)

****

# ä»–ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚‚CWLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦è¨˜è¿°ã™ã‚‹

blastpã®ãƒ—ãƒ­ã‚»ã‚¹ã®ã‚ˆã†ã«ï½¤ä»–ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚‚CWLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦æ›¸ã„ã¦ã¿ã¾ã™ï½¡
__zatsu-cwl-generatorã‚’ä½¿ã£ã¦ç”Ÿæˆã—__ï½¤ä¿®æ­£ã—ãŸãƒãƒ¼ã‚¸ãƒ§ãƒ³ã¯`v2`ãªã©ã®ã‚ˆã†ã«ãƒ•ã‚¡ã‚¤ãƒ«åã‚’å¤‰æ›´ã—ã¦ã„ã¾ã™ï½¡
ã¾ãŸï½¤`awk`ä»¥å¤–ã®ãƒ—ãƒ­ã‚»ã‚¹ã¯docker imageã¨ã—ã¦BioContainersã«ã‚ã‚‹ã‚‚ã®ã‚’ä½¿ç”¨ã—ï½¤`hints`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’è¿½åŠ ã—ã¦ã„ã¾ã™ï½¡

:::message
ä»Šå›å–ã‚Šä¸Šã’ã¦ã„ã‚‹ãƒ„ãƒ¼ãƒ«ã®docker imageã‚’[docker hub](https://hub.docker.com/)ã‹ã‚‰å–å¾—ã—ã¾ã—ãŸï½¡
[blast](https://hub.docker.com/r/biocontainers/blast)ï½¤[clustalo](https://hub.docker.com/r/biocontainers/clustalo)ï½¤[fasttree](https://hub.docker.com/r/biocontainers/fasttree)ã¯ã™ã¹ã¦[BioContainers](https://biocontainers.pro/)ã«ã‚ˆã£ã¦æ§‹ç¯‰ã•ã‚Œã¦ã„ã¾ã™ï½¡
```bash:
docker image ls
REPOSITORY               TAG                 IMAGE ID       CREATED       SIZE
biocontainers/fasttree   v2.1.10-2-deb_cv1   bdfde5d6026c   4 years ago   118MB
biocontainers/clustalo   v1.2.4-2-deb_cv1    d14a5e27e301   4 years ago   118MB
biocontainers/blast      v2.2.31_cv2         5b25e08b9871   5 years ago   2.03GB
```
:::

:::details 2_awk_v2.cwl
### (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦ç”Ÿæˆ

```bash
zatsu-cwl-generator "awk '{ print $2 }' blastp_result.txt > blastp_result_id.txt" > 2_awk.cwl
```
awkã®ãƒ—ãƒ­ã‚»ã‚¹ã§ã¯ï½¤`arguments`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’è¿½åŠ ã—ï½¤ä»¥ä¸‹ã®ã‚ˆã†ã«ä¿®æ­£ã—ã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/2_awk_v2.cwl

### (2) CWLãƒ•ã‚¡ã‚¤ãƒ«ã®è©•ä¾¡

```bash
cwltool --validate 2_awk_v2.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '2_awk_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/2_awk_v2.cwl'
2_awk_v2.cwl is valid CWL.
```

### (3) å®Ÿè¡Œ

```bash:å®Ÿè¡Œ
cwltool --debug 2_awk_v2.cwl
```
å®Ÿè¡Œã—ãŸãƒ•ã‚¡ã‚¤ãƒ«ã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blastp_result_id.txt
:::

:::details 3_blastdbcmd_v2.cwl
### (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦ç”Ÿæˆ

```bash
zatsu-cwl-generator "blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastdbcmd_result.fasta" --container biocontainers/blast:v2.2.31_cv2 > 3_blastdbcmd_docker.cwl
```

`1_blastp.cwl`ã®ã‚ˆã†ã«ï½¤`secondaryFiles`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ä½¿ã£ã¦è¤‡æ•°ã®ã‚¤ãƒ³ãƒ‡ãƒƒã‚¯ã‚¹ãƒ•ã‚¡ã‚¤ãƒ«ã‚’æŒ‡å®šã—ã¦ã„ã‚‹ä»–ï½¤ãƒ•ã‚¡ã‚¤ãƒ«åã‚’`string`ã«å¤‰æ›´ã—ï½¤ä¿®æ­£ã‚’åŠ ãˆã¾ã—ãŸï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/3_blastdbcmd_docker_v2.cwl

### (2) CWLãƒ•ã‚¡ã‚¤ãƒ«ã®è©•ä¾¡

```bash:
cwltool --validate 3_blastdbcmd_docker_v2.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '3_blastdbcmd_docker_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/3_blastdbcmd_docker_v2.cwl'
3_blastdbcmd_docker_v2.cwl is valid CWL.
```

### (3) å®Ÿè¡Œ

```bash:å®Ÿè¡Œ
cwltool --debug 3_blastdbcmd_docker_v2.cwl
```

å®Ÿè¡Œã—ãŸçµæœã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡ ç„¡äº‹ï½¤ãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blastdbcmd_result.fasta
:::

:::details 4_clustalo.cwl
### (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦ç”Ÿæˆ

```bash
zatsu-cwl-generator "clustalo -i blastdbcmd_result.fasta --outfmt=fasta -o clustalo_result.fasta" --container biocontainers/clustalo:v1.2.4-2-deb_cv1 > 4_clustalo_docker.cwl
```

### (2) CWLãƒ•ã‚¡ã‚¤ãƒ«ã®è©•ä¾¡

```bash:
cwltool --validate 4_clustalo_docker.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '4_clustalo_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/4_clustalo_docker.cwl'
4_clustalo_docker.cwl is valid CWL.
```

clustaloã®ãƒ—ãƒ­ã‚»ã‚¹ã§ã¯ï½¤__ä¿®æ­£ãŒå¿…è¦ãªãï½¤ç”Ÿæˆã™ã‚‹ã“ã¨ãŒã§ãã¦ã„ã‚‹ã‚ˆã†ã§ã™ï½¡__ (ã¨æ€ã„ã¾ã—ãŸãŒï½¤ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã®è¨˜è¿°ã®éš›ã«ä¿®æ­£ãŒå¿…è¦ã«ãªã‚Šã¾ã—ãŸ)

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/4_clustalo_docker.cwl

### (3) å®Ÿè¡Œ

```bash:å®Ÿè¡Œ
cwltool --debug 4_clustalo_docker.cwl
```

å®Ÿè¡Œã—ãŸçµæœã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡ ç„¡äº‹ï½¤ãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/clustalo_result.fasta
:::

:::details 5_fasttree.cwl
### (1) zatsu-cwl-generatorã‚’ä½¿ã£ã¦ç”Ÿæˆ

```bash
zatsu-cwl-generator "fasttree -nt clustalo_result.fasta > fasttree_result.nwk" --container biocontainers/fasttree:v2.1.10-2-deb_cv1 > 5_fasttree_docker.cwl
```

### (2) CWLãƒ•ã‚¡ã‚¤ãƒ«ã®è©•ä¾¡

```bash:
cwltool --validate 5_fasttree_docker.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved '5_fasttree_docker.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/5_fasttree_docker.cwl'
5_fasttree_docker.cwl is valid CWL.
```

fasttreeã®ãƒ—ãƒ­ã‚»ã‚¹ã§ã¯ï½¤__ä¿®æ­£ãŒå¿…è¦ãªãï½¤ç”Ÿæˆã™ã‚‹ã“ã¨ãŒã§ãã¦ã„ã‚‹ã‚ˆã†ã§ã™ï½¡__

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/5_fasttree_docker.cwl

### (3) å®Ÿè¡Œ

```bash:å®Ÿè¡Œ
cwltool --debug 5_fasttree_docker.cwl
```

å®Ÿè¡Œã—ãŸçµæœã¯ä»¥ä¸‹ã®é€šã‚Šã§ã™ï½¡ ç„¡äº‹ï½¤ãƒ•ã‚¡ã‚¤ãƒ«ãŒå‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/fasttree_result.nwk
:::

å…¨ã¦ã®ãƒ—ãƒ­ã‚»ã‚¹ã‚’CWLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦æ›¸ãã“ã¨ãŒã§ãã¾ã—ãŸï½¡å°‘ã—ä¿®æ­£ã™ã‚‹éƒ¨åˆ†ãŒå¿…è¦ã§ã—ãŸãŒï½¤ãŠãã‚‰ãã“ã‚Œã§ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚‚æ›¸ã‘ã‚‹ã¯ãšã§ã™ï½¡
__ã“ã®ã‚ˆã†ã«ï½¤zatsu-cwl-generatorã‚’ä½¿ã†ã“ã¨ã§ï½¤ç°¡å˜ã«CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡__

:::message
### Apptainer (Singularity)ã‚‚ä½¿ã£ã¦ã¿ã‚ˆã†

dockerã®ã‚³ãƒ³ãƒ†ãƒŠã¯ï½¤`root`æ¨©é™ã‚’æŒã¤ãƒ¦ãƒ¼ã‚¶ãƒ¼ãŒå®Ÿè¡Œã—ã¾ã™ï½¡
ã—ã‹ã—ãªãŒã‚‰ï½¤å›½ç«‹éºä¼å­¦ç ”ç©¶æ‰€ãŒæ‰€æœ‰ã—ã¦ã„ã‚‹ã‚¹ãƒ¼ãƒ‘ãƒ¼ã‚³ãƒ³ãƒ”ãƒ¥ãƒ¼ã‚¿ãƒ¼ã®ã‚ˆã†ãªç’°å¢ƒã§ã¯ï½¤`root`æ¨©é™ã‚’æŒã¤ãƒ¦ãƒ¼ã‚¶ãƒ¼ãŒå®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã›ã‚“ï½¡
ãã“ã§ï½¤`Singularity`ã‚’ä½¿ã£ã¦ã‚³ãƒ³ãƒ†ãƒŠã‚’å®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
ä¾‹ãˆã°ï½¤docker hubã§æ¤œç´¢ã—ãŸã‚³ãƒ³ãƒ†ãƒŠã‚’å¤‰æ›ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
:::

https://sc.ddbj.nig.ac.jp/#%E9%81%BA%E4%BC%9D%E7%A0%94%E3%82%B9%E3%83%BC%E3%83%91%E3%83%BC%E3%82%B3%E3%83%B3%E3%83%94%E3%83%A5%E3%83%BC%E3%82%BF%E3%82%B7%E3%82%B9%E3%83%86%E3%83%A0

https://sc.ddbj.nig.ac.jp/software/Apptainer/

****

# ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’è¨˜è¿°ã™ã‚‹

ã“ã‚Œã¾ã§ï½¤`Commandlinetool`ã®CWLãƒ•ã‚¡ã‚¤ãƒ«(ä¸€ã¤ä¸€ã¤ã®å‡¦ç†ã‚’è¨˜è¿°)ã‚’æ›¸ãã¾ã—ãŸï½¡
æ¬¡ã«ã“ã‚Œã‚‰ã®5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’å®Ÿè¡Œã™ã‚‹ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ï½¤`blast2tree.cwl`ã‚’è¨˜è¿°ã—ã¦ã„ãã¾ã™ï½¡
ã“ã®ä¾‹ã§ã¯ï½¤ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼å…¨ä½“ã«é–¢ã™ã‚‹ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’ï½¢1\_protein\_queryï½£ã®ã‚ˆã†ã«æ•°å­—(`1_`ï½¤`2_`)ã‚’ã¤ã‘ã¦ã„ã¾ã™ï½¡ 
å‰ã®å‡¦ç†ã®ã‚¢ã‚¦ãƒˆãƒ—ãƒƒãƒˆã‚’å—ã‘å–ã‚‹éƒ¨åˆ†ã¯ï½¢blastp\_result: step1\_blastp/blastp\_output\_fileï½£ã®ã‚ˆã†ã«è¨˜è¿°ã—ï½¤ãã‚Œä»¥å¤–ã®å…¨ã¦ã®å‡¦ç†ã«é–¢ã™ã‚‹ãƒ‘ãƒ©ãƒ¡ãƒ¼ã‚¿ã‚’inputsã§è¨˜è¿°ã—ã¦ã„ã¾ã™ï½¡

&nbsp;

ä½œæˆã—ãŸCWLãƒ•ã‚¡ã‚¤ãƒ«ã¯ä»¥ä¸‹ã®ã‚ˆã†ã«ãªã£ã¦ã„ã¾ã™ï½¡`inputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ï½¤`steps`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ï½¤`outputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ãã‚Œãã‚Œæ›¸ã„ã¦ã¿ã¾ã—ãŸï½¡ 
ã¾ãŸï½¤å¾Œã«å®Ÿè¡Œã™ã‚‹éš›ã®ãŸã‚ã«ï½¤`doc`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚‚è¿½åŠ ã—ã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blast2tree.cwl

## æ­£ã—ã„è¨˜è¿°ã‹ã©ã†ã‹ã‚’ç¢ºèªã™ã‚‹

ä¸€æ—¦ `cwltool --validate`ã§æ›¸ãæ–¹ã‚’ç¢ºèªã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

```bash
cwltool --validate blast2tree.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved 'blast2tree.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blast2tree.cwl'
ERROR Tool definition failed validation:

blast2tree.cwl:99:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:104:7:   with sink 'nt' of type "File"
blast2tree.cwl:79:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:84:7:     with sink 'blastp_result' of type "File"
blast2tree.cwl:92:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:97:7:     with sink 'i' of type "File"
blast2tree.cwl:92:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:117:5:   with sink 'blastdbcmd_output' of type "File"
blast2tree.cwl:79:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:109:5:   with sink 'blastp_output' of type "File"
blast2tree.cwl:99:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File",
                      "Directory"]} is incompatible
blast2tree.cwl:121:5:   with sink 'clustalo_output' of type "File"
```

ãªã«ã‚„ã‚‰ã‚¨ãƒ©ãƒ¼ãŒå‡ºã¦ã„ã¾ã™ï½¡ `blast2tree.cwl:79:11: Source 'all-for-debugging' of type {"type": "array", "items": ["File","Directory"]} is incompatible` ã¨ã„ã†ã‚ˆã†ã«ï½¤å‹ã®ä¸ä¸€è‡´ãŒã‚ã‚‹ã‚ˆã†ã§ã™ï½¡
ã“ã®ã‚¨ãƒ©ãƒ¼ã‚’è§£æ¶ˆã™ã‚‹ãŸã‚ã«ï½¤`CommandLineTool`å®šç¾©ã®ãã‚Œãã‚Œ5ã¤ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’å°‘ã—ä¿®æ­£ã—ã¦ã¿ã¾ã™ï½¡
ä¾‹ãˆã°ï½¤`1_blastp_docker_v4.cwl`ã®`outputs`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ä»¥ä¸‹ã®ã‚ˆã†ã«ä¿®æ­£ã—ã¦ã¿ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/1_blastp_docker_v5.cwl#L56-L58

ä»–ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚‚åŒæ§˜ã«å°‘ã—ä¿®æ­£ã—ã¾ã™ï½¡(ãƒãƒ¼ã‚¸ãƒ§ãƒ³ã‚’ã‚ã’ã¾ã—ãŸ)

:::details ä¿®æ­£ã—ãŸãƒ•ã‚¡ã‚¤ãƒ«
https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/3_blastdbcmd_docker_v3.cwl

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/4_clustalo_docker_v2.cwl
:::

å†åº¦ä½œæˆã—ãŸCWLãƒ•ã‚¡ã‚¤ãƒ«ã¯ä»¥ä¸‹ã®ã‚ˆã†ã«ãªã£ã¦ã„ã¾ã™ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blast2tree_v2.cwl

ã“ã‚Œã§å†åº¦`cwltool --validate`ã‚’å®Ÿè¡Œã—ã¦ã¿ã¾ã™ï½¡

```bash
cwltool --validate blast2tree_v2.cwl
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved 'blast2tree_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blast2tree_v2.cwl'
blast2tree_v2.cwl is valid CWL.
```
ã“ã‚Œã§ã‚¨ãƒ©ãƒ¼ãŒè§£æ¶ˆã•ã‚Œã¾ã—ãŸ! 

## CWLviewerã§ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’ç¢ºèªã™ã‚‹

ã“ã®ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã®å…¨ä½“åƒã‚’CWLviewerã§ç¢ºèªã—ã¾ã—ã‚‡ã†ï½¡
[`blast2tree_v2.cwl`](https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blast2tree_v2.cwl)ã‚’CWLviewerã«ã‚¢ãƒƒãƒ—ãƒ­ãƒ¼ãƒ‰ã—ã¦ã¿ã¾ã™ï½¡

https://view.commonwl.org/workflows/github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/blast2tree_v2.cwl

![CWLviewer3](https://storage.googleapis.com/zenn-user-upload/f4574f7c2bce-20240612.png)

å…¨ä½“ã¨ã—ã¦ã¯ã“ã®ã‚ˆã†ã«ãªã£ã¦ã„ã¾ã™ï½¡5ã¤ã‚‚ã®ã‚¹ãƒ†ãƒƒãƒ—ãŒãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¨ã—ã¦è¨˜è¼‰ã•ã‚Œã¦ã„ã‚‹ã“ã¨ãŒæ”¹ã‚ã¦ã‚ã‹ã‚Šã¾ã™ï½¡
(æ¯å›è‡ªåˆ†ã¯ã“ã‚Œã«æ„Ÿå‹•ã—ã¦ã„ã¾ã™)

## å®Ÿéš›ã«å®Ÿè¡Œã™ã‚‹

ãã‚Œã§ã¯å®Ÿéš›ã«å®Ÿè¡Œã—ã¦ã¿ã¾ã—ã‚‡ã†ï½¡
ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã®çµæœã‚’åˆ†ã‘ã‚‹ãŸã‚ã«ï½¤`--outdir`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ä½¿ã£ã¦å®Ÿè¡Œã—ã¾ã™ï½¡

```bash
cwl-runner --outdir ./workflow_result/ blast2tree_v2.cwl
```
:::details ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼å®Ÿè¡Œãƒ—ãƒ­ã‚»ã‚¹

```bash
cwl-runner --outdir ./workflow_result/ blast2tree_v2.cwl
INFO /usr/local/bin/cwl-runner 3.1.20240508115724
INFO Resolved 'blast2tree_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blast2tree_v2.cwl'
INFO [workflow ] start
INFO [workflow ] starting step step1_blastp
INFO [step step1_blastp] start
INFO [job step1_blastp] /tmp/uj4w3esx$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/uj4w3esx,target=/WlXEKf \
    --mount=type=bind,source=/tmp/pozju8zg,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/MSTN.fasta,target=/var/lib/cwl/stgced530ce-9349-463a-99b6-aa9c4fc9f28c/
MSTN.fasta,readonly \                                                                                                                                                        --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-2d3c
6f883a16/uniprot_sprot.fasta,readonly \                                                                                                                                      --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.phd,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.phi,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.phr,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.pin,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.pog,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.psd,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.psi,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq,target=/var/lib/cwl/stg0f1360dc-539d-4d47-99f0-
2d3c6f883a16/uniprot_sprot.fasta.psq,readonly \                                                                                                                              --workdir=/WlXEKf \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/n2b6wz52/20240612103142-461008.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/WlXEKf \
    biocontainers/blast:v2.2.31_cv2 \
    blastp \
    -query \
    /var/lib/cwl/stgced530ce-9349-463a-99b6-aa9c4fc9f28c/MSTN.fasta \
    -db \
    /var/lib/cwl/stg0f1360dc-539d-4d47-99f0-2d3c6f883a16/uniprot_sprot.fasta \
    -evalue \
    0.01 \
    -num_threads \
    8 \
    -outfmt \
    6 \
    -out \
    blastp_output.txt \
    -max_target_seqs \
    20
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
INFO [job step1_blastp] Max memory used: 0MiB
INFO [job step1_blastp] completed success
INFO [step step1_blastp] completed success
INFO [workflow ] starting step step2_awk
INFO [step step2_awk] start
INFO [job step2_awk] /tmp/6lqwjots$ awk \
    '{ print $2 }' \
    /tmp/4qdjlcga/stge55ffb48-5bb9-454a-a7ff-c429efc74aaf/blastp_output.txt > /tmp/6lqwjots/blastp_result_id.txt
INFO [job step2_awk] completed success
INFO [step step2_awk] completed success
INFO [workflow ] starting step step3_blastdbcmd
INFO [step step3_blastdbcmd] start
INFO [job step3_blastdbcmd] /tmp/rdv3hv2x$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/rdv3hv2x,target=/WlXEKf \
    --mount=type=bind,source=/tmp/r1olsbb3,target=/tmp \
    --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-a5fa
4f0738bb/uniprot_sprot.fasta,readonly \                                                                                                                                      --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phd,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.phd,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phi,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.phi,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.phr,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.phr,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pin,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.pin,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.pog,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.pog,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psd,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.psd,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psi,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.psi,readonly \                                                                                                                              --mount=type=bind,source=/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/uniprot_sprot.fasta.psq,target=/var/lib/cwl/stgc0263af5-6318-4781-8db1-
a5fa4f0738bb/uniprot_sprot.fasta.psq,readonly \                                                                                                                              --mount=type=bind,source=/tmp/6lqwjots/blastp_result_id.txt,target=/var/lib/cwl/stg3930e103-780f-4d8c-bdca-c36a8e0359e3/blastp_result_id.txt,readonly \
    --workdir=/WlXEKf \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/4hh0ljvr/20240612103143-671398.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/WlXEKf \
    biocontainers/blast:v2.2.31_cv2 \
    blastdbcmd \
    -db \
    /var/lib/cwl/stgc0263af5-6318-4781-8db1-a5fa4f0738bb/uniprot_sprot.fasta \
    -entry_batch \
    /var/lib/cwl/stg3930e103-780f-4d8c-bdca-c36a8e0359e3/blastp_result_id.txt \
    -out \
    blastdbcmd_result.fasta
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
INFO [job step3_blastdbcmd] completed success
INFO [step step3_blastdbcmd] completed success
INFO [workflow ] starting step step4_clustalo
INFO [step step4_clustalo] start
INFO [job step4_clustalo] /tmp/xz4llstr$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/xz4llstr,target=/WlXEKf \
    --mount=type=bind,source=/tmp/c8axor6c,target=/tmp \
    --mount=type=bind,source=/tmp/rdv3hv2x/blastdbcmd_result.fasta,target=/var/lib/cwl/stg8426b713-ef9d-46bc-8e36-daaa52a1f2f8/blastdbcmd_result.fasta,readonly \
    --workdir=/WlXEKf \
    --read-only=true \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/6ykuaryi/20240612103144-728058.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/WlXEKf \
    biocontainers/clustalo:v1.2.4-2-deb_cv1 \
    clustalo \
    -i \
    /var/lib/cwl/stg8426b713-ef9d-46bc-8e36-daaa52a1f2f8/blastdbcmd_result.fasta \
    --outfmt=fasta \
    -o \
    clustalo_output.fasta
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
INFO [job step4_clustalo] completed success
INFO [step step4_clustalo] completed success
INFO [workflow ] starting step step5_fasttree
INFO [step step5_fasttree] start
INFO [job step5_fasttree] /tmp/9_v2dbe8$ docker \
    run \
    -i \
    --mount=type=bind,source=/tmp/9_v2dbe8,target=/WlXEKf \
    --mount=type=bind,source=/tmp/k186douq,target=/tmp \
    --mount=type=bind,source=/tmp/xz4llstr/clustalo_output.fasta,target=/var/lib/cwl/stgb1ab0516-de4a-4a6c-b9ad-47f48f617aa4/clustalo_output.fasta,readonly \
    --workdir=/WlXEKf \
    --read-only=true \
    --log-driver=none \
    --user=1000:1000 \
    --rm \
    --cidfile=/tmp/9an6c2tv/20240612103145-770529.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/WlXEKf \
    biocontainers/fasttree:v2.1.10-2-deb_cv1 \
    fasttree \
    -nt \
    /var/lib/cwl/stgb1ab0516-de4a-4a6c-b9ad-47f48f617aa4/clustalo_output.fasta > /tmp/9_v2dbe8/fasttree_result.nwk
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
FastTree Version 2.1.10 Double precision (No SSE3)
Alignment: /var/lib/cwl/stgb1ab0516-de4a-4a6c-b9ad-47f48f617aa4/clustalo_output.fasta
Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000
Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1
TopHits: 1.00*sqrtN close=default refresh=0.80
ML Model: Jukes-Cantor, CAT approximation with 20 rate categories
Ignored unknown character D (seen 311 times)
Ignored unknown character E (seen 310 times)
Ignored unknown character F (seen 176 times)
Ignored unknown character H (seen 77 times)
Ignored unknown character I (seen 332 times)
Ignored unknown character K (seen 364 times)
Ignored unknown character L (seen 451 times)
Ignored unknown character M (seen 110 times)
Ignored unknown character P (seen 312 times)
Ignored unknown character Q (seen 232 times)
Ignored unknown character R (seen 237 times)
Ignored unknown character S (seen 278 times)
Ignored unknown character V (seen 279 times)
Ignored unknown character W (seen 78 times)
Ignored unknown character X (seen 205 times)
Ignored unknown character Y (seen 169 times)
WARNING! ONLY 19.6% NUCLEOTIDE CHARACTERS -- IS THIS REALLY A NUCLEOTIDE ALIGNMENT?
Initial topology in 0.01 seconds
Refining topology: 15 rounds ME-NNIs, 2 rounds ME-SPRs, 7 rounds ML-NNIs
Total branch-length 0.107 after 0.01 sec
ML-NNI round 1: LogLk = -167.434 NNIs 5 max delta 0.00 Time 0.02
Switched to using 20 rate categories (CAT approximation)
Rate categories were divided by 0.628 so that average rate = 1.0
CAT-based log-likelihoods may not be comparable across runs
Use -gamma for approximate but comparable Gamma(20) log-likelihoods
ML-NNI round 2: LogLk = -165.205 NNIs 3 max delta 0.00 Time 0.02
Turning off heuristics for final round of ML NNIs (converged)
ML-NNI round 3: LogLk = -165.204 NNIs 2 max delta 0.00 Time 0.03 (final)
Optimize all lengths: LogLk = -165.204 Time 0.03
Total time: 0.06 seconds Unique: 13/20 Bad splits: 0/10
INFO [job step5_fasttree] completed success
INFO [step step5_fasttree] completed success
INFO [workflow ] completed success
{
    "awk_output": {
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastp_result_id.txt",
        "basename": "blastp_result_id.txt",
        "class": "File",
        "checksum": "sha1$354af893fc9afa558d0a01d6d68bf9d3a935cb53",
        "size": 418,
        "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastp_result_id.txt"
    },
    "blastdbcmd_output": {
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastdbcmd_result.fasta",
        "basename": "blastdbcmd_result.fasta",
        "class": "File",
        "checksum": "sha1$e9167fd8855ce938590f64e179b3aedb0edd9e15",
        "size": 9581,
        "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastdbcmd_result.fasta"
    },
    "blastp_output": {
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastp_output.txt",
        "basename": "blastp_output.txt",
        "class": "File",
        "checksum": "sha1$31464e2d0bdfd65f89e0542688844d3686bf278c",
        "size": 1536,
        "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/blastp_output.txt"
    },
    "clustalo_output": {
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/clustalo_output.fasta",
        "basename": "clustalo_output.fasta",
        "class": "File",
        "checksum": "sha1$57ec8d531af2e6a5a0add1f1fc8ff413a12d62f6",
        "size": 9621,
        "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/clustalo_output.fasta"
    },
    "fasttree_output": {
        "location": "file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/fasttree_result.nwk",
        "basename": "fasttree_result.nwk",
        "class": "File",
        "checksum": "sha1$90a12b35e23309a1e73fac58d05794beeab49a38",
        "size": 807,
        "path": "/workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/workflow_result/fasttree_result.nwk"
    }
}INFO Final process status is success
```
:::

ç„¡äº‹ï½¤å…¨ã¦ã®å‡¦ç†ãŒçµ‚äº†ã—ã¦ã„ã‚‹ã‚ˆã†ã§ã™! ãã‚Œã§ã¯ï½¤å®Ÿéš›ã«å‡ºåŠ›ã•ã‚ŒãŸãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¦‹ã¦ã¿ã¾ã—ã‚‡ã†ï½¡

https://github.com/yonesora56/togotv_cwl_for_remote_container/blob/master/zatsu_cwl_bioinformatics/workflow_result/fasttree_result.nwk

ç„¡äº‹ï½¤å‡ºåŠ›ã•ã‚Œã¦ã„ã¾ã™ï½¡ã“ã®ã‚ˆã†ã«ï½¤ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã‚’CWLã§è¨˜è¿°ã™ã‚‹ã“ã¨ã§ï½¤ä¸€ã¤ã®ã‚³ãƒãƒ³ãƒ‰ã§è¤‡æ•°ã®å‡¦ç†ã‚’ä¸€åº¦ã«å®Ÿè¡Œã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡

&nbsp;

:::message
### `doc`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ç©æ¥µçš„ã«æ›¸ã“ã†!

è¤‡é›‘ãªãƒ—ãƒ­ã‚»ã‚¹ã‚’æ›¸ã„ãŸã‚ã¨ï½¤è¦‹è¿”ã—ãŸéš›ã«ï½¢ã‚ã‚Œï½¤ã“ã®ãƒ—ãƒ­ã‚»ã‚¹ã¯ãªã‚“ã ã£ã‘...ï½£ã¨ãªã£ã¦ã—ã¾ã†ã“ã¨ã¯å®¹æ˜“ã«æƒ³åƒã§ãã¾ã™ï½¡
ãã“ã§ï½¤CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’æ›¸ãéš›ã«ã¯ï½¤`doc`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’ç©æ¥µçš„ã«æ›¸ãã“ã¨ã‚’ãŠã™ã™ã‚ã—ã¾ã™ï½¡

ä¾‹ãˆã°ï½¤å…ˆç¨‹ä½œæˆã—ãŸ`Workflow`å®šç¾©ã®CWLãƒ•ã‚¡ã‚¤ãƒ«ã«å¯¾ã—ã¦`--help`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã¦å®Ÿè¡Œã—ã¦ã¿ã¾ã™(ä½ç½®ã«æ³¨æ„!)ï½¡

```bash:
cwltool blast2tree_v2.cwl --help
INFO /usr/local/bin/cwltool 3.1.20240508115724
INFO Resolved 'blast2tree_v2.cwl' to 'file:///workspaces/togotv_cwl_for_remote_container/zatsu_cwl_bioinformatics/blast2tree_v2.cwl'
usage: blast2tree_v2.cwl [-h] [--1_protein_query 1_PROTEIN_QUERY] [--2_protein_database 2_PROTEIN_DATABASE] [--3_evalue 3_EVALUE]
                         [--4_number_of_threads 4_NUMBER_OF_THREADS] [--5_outformat_type 5_OUTFORMAT_TYPE] [--6_output_file_name 6_OUTPUT_FILE_NAME]
                         [--7_max_target_sequence 7_MAX_TARGET_SEQUENCE] [--8_blastdbcmd_protein_database 8_BLASTDBCMD_PROTEIN_DATABASE]
                         [--10_clustalo_output_name 10_CLUSTALO_OUTPUT_NAME]
                         [job_order]

blastp, awk, blastdbcmd, clustalo, and fasttreeã®5ã¤ã®ã‚¹ãƒ†ãƒƒãƒ—ã‚’å®Ÿè¡Œ

positional arguments:
  job_order             Job input json file

options:
  -h, --help            show this help message and exit
  --1_protein_query 1_PROTEIN_QUERY
                        blastp process input protein query file(fasta format)
  --2_protein_database 2_PROTEIN_DATABASE
                        blastp process input all index file
  --3_evalue 3_EVALUE   blastp process evalue cutoff
  --4_number_of_threads 4_NUMBER_OF_THREADS
                        number of threads
  --5_outformat_type 5_OUTFORMAT_TYPE
                        blastp process output format type
  --6_output_file_name 6_OUTPUT_FILE_NAME
                        blastp process output file name
  --7_max_target_sequence 7_MAX_TARGET_SEQUENCE
                        blastp process max target sequence
  --8_blastdbcmd_protein_database 8_BLASTDBCMD_PROTEIN_DATABASE
                        blastdbcmd process protein database
  --10_clustalo_output_name 10_CLUSTALO_OUTPUT_NAME
                        clustalo process output name
```
ã“ã®ã‚ˆã†ã«ï½¤`doc`ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰ã‚’æ›¸ã„ã¦ãŠãã“ã¨ã§ï½¤ãƒ—ãƒ­ã‚»ã‚¹ã®èª¬æ˜ã‚’è¦‹ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
:::

# çµ‚ã‚ã‚Šã«

ä»Šå›ã¯ï½¤ãƒã‚¤ã‚ªã‚¤ãƒ³ãƒ•ã‚©ãƒãƒ†ã‚£ã‚¯ã‚¹åˆ†é‡ã§ä½¿ç”¨ã•ã‚Œã‚‹ãƒ„ãƒ¼ãƒ«ã‚’CWLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦è¨˜è¿°ã—ï½¤ãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼ã¾ã§å®Ÿè¡Œã™ã‚‹ãƒ—ãƒ­ã‚»ã‚¹ã‚’è¨˜è¿°ã—ã¾ã—ãŸï½¡
zatsu-cwl-generatorã‚’ä½¿ã†ã“ã¨ã§ï½¤ç°¡å˜ã«CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ç”Ÿæˆã™ã‚‹ã“ã¨ã§ï½¤CWLã‚’è¨˜è¿°ã™ã‚‹ãƒãƒ¼ãƒ‰ãƒ«ã‚’ä¸‹ã’ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡
ã“ã‚Œã¾ã§ã®è¨˜äº‹ã§èª¬æ˜ã—ã¦ã„ã‚‹ãƒ—ãƒ­ã‚»ã‚¹ã¨ã—ã¦ã¯ï½¤ã¾ã¨ã‚ã‚‹ã¨ä»¥ä¸‹ã®ã‚ˆã†ã«ãªã£ã¦ã„ã¾ã™ï½¡

1. zatsu-cwl-generatorã‚’ä½¿ã£ã¦ï½¤ã‚³ãƒãƒ³ãƒ‰ãƒ©ã‚¤ãƒ³ãƒ„ãƒ¼ãƒ«ã®å‡¦ç†ã‚’CWLãƒ•ã‚¡ã‚¤ãƒ«ã¨ã—ã¦ç”Ÿæˆã™ã‚‹
2. cwltoolã® `--validate`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ä½¿ã£ã¦ï½¤CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãƒã‚§ãƒƒã‚¯ã™ã‚‹
3. å•é¡ŒãŒã‚ã‚‹å ´åˆï½¤ä¿®æ­£ã‚’è¡Œã†(ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã‚’èª­ã‚€ï½¤ãªã©)
4. cwltoolã‚’ä½¿ã£ã¦ï½¤CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’å®Ÿè¡Œã™ã‚‹(`--debug`ã‚ªãƒ—ã‚·ãƒ§ãƒ³ã‚’ã¤ã‘ã¦ãŠã)
5. ã‚¨ãƒ©ãƒ¼ãŒå‡ºãŸå ´åˆï½¤ä¿®æ­£ã‚’è¡Œã†

ã“ã®ã‚ˆã†ã«ï½¤ __zatsu-cwl-generator__ ã‚’CWLãƒ•ã‚¡ã‚¤ãƒ«ã‚’æ›¸ããƒ—ãƒ­ã‚»ã‚¹ã®èµ·ç‚¹ã¨ã—ã¦åˆ©ç”¨ã™ã‚‹ã“ã¨ãŒå¯èƒ½ã§ã™ï½¡
ãœã²ï½¤çš†ã•ã‚“ã‚‚æ´»ç”¨ã—ã¦ç´ æ™´ã‚‰ã—ã„CWLãƒ©ã‚¤ãƒ•ã‚’æº€å–«ã—ã¦ãã ã•ã„!

&nbsp;

## å‚è€ƒ

ã“ã‚Œã¾ã§ã®3ã¤ã®è¨˜äº‹ã§ã¯è©³ã—ã„éƒ¨åˆ†ã«ã¤ã„ã¦ã¯ã‚ã¾ã‚Šè§£èª¬ã—ã¦ã„ã¾ã›ã‚“ï½¡
CWLã®è©³ç´°ãªè¨˜è¿°æ–¹æ³•ã«ã¤ã„ã¦ã¯ï½¤ä»¥ä¸‹ã®ã‚ˆã†ãªãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆãŒå……å®Ÿã—ã¦ã„ã‚‹ã®ã§ï½¤ãœã²ãã¡ã‚‰ã‚’å‚è€ƒã«ã—ã¦ãã ã•ã„ï½¡

https://www.commonwl.org/user_guide/#common-workflow-language-user-guide

https://www.commonwl.org/user_guide/topics/best-practices.html#best-practices

&nbsp;

:::details å‚è€ƒï¼šDockerã§ã‚³ãƒãƒ³ãƒ‰å®Ÿè¡Œã™ã‚‹ä¾‹
```bash:
# Step1
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/blast:v2.2.31_cv2 blastp -query MSTN.fasta -db uniprot_sprot.fasta -evalue 1e-5 -num_threads 4 -outfmt 6 -out blastp_result_MSTN.txt -max_target_seqs 20 

# Step2
docker run --rm -it -v `pwd`:`pwd` -w `pwd` ubuntu:23.10 awk '{ print $2 }' `pwd`/data/blastp_result.txt > `pwd`/data/blastp_result_id.txt

#Step3
docker run --rm -it -v `pwd`:`pwd`  -w  `pwd` biocontainers/blast:v2.2.31_cv2 blastdbcmd -db uniprot_sprot.fasta -entry_batch blastp_result_id.txt  -out blastdbcmd_result.fasta

#Step4
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/clustalo:v1.2.4-2-deb_cv1 clustalo -i `pwd`/blastp_results_MSTN.fasta --outfmt=fasta -o clustalo_result.fasta

#Step5
docker run --rm -it -v `pwd`:`pwd` -w `pwd` biocontainers/fasttree:v2.1.10-2-deb_cv1 FastTree -boot 100 -out MSTN_tree.newick `pwd`/clustalo_result.fasta
```
:::

&nbsp;

__ä»Šå›ã®è¨˜äº‹ã§ä½¿ç”¨ã—ãŸCWLã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ãŠã„ã¦ã„ã‚‹ãƒªãƒã‚¸ãƒˆãƒªã¯ä»¥ä¸‹ã‹ã‚‰ã‚¢ã‚¯ã‚»ã‚¹ã™ã‚‹ã“ã¨ãŒã§ãã¾ã™ï½¡__
https://github.com/yonesora56/togotv_cwl_for_remote_container